---
title: "c9alsftd_multiome - differential comparison"
author: "Hsiao-Lin Veronica Wang | Corces Lab | Department of Human Genetics | Emory University School of Medicine"
date: "`r Sys.Date()`"
output:
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(eval = FALSE)
```


## Load required packages and graphical set up

```{r}
library(ArchR)
library(parallel)
library(monocle3)
library(tidyselect)
library(tidygraph)
library(tidytree
library(tidyr
library(tidyverse)
library(parallel)
library(monocle3)
library(cicero)

# https://rpubs.com/Koundy/71792
# remove all bold text
theme_Publication <- function(base_size=7, base_family="sans") {
      library(grid)
      library(ggthemes)
      (theme_foundation(base_size=base_size, base_family=base_family)
       + theme(plot.title = element_text(face = "bold",
                                         size = rel(1.2), hjust = 0.5),
               text = element_text(),
               panel.background = element_rect(colour = NA),
               plot.background = element_rect(colour = NA),
               panel.border = element_rect(colour = NA),
               axis.title = element_text(size = rel(1)),
               axis.title.y = element_text(angle=90,vjust =2),
               axis.title.x = element_text(vjust = -0.2),
               axis.text = element_text(), 
               axis.line = element_line(colour="black", size = 0.2),
               axis.ticks = element_line(),
               panel.grid.major = element_line(colour="#f0f0f0"),
               panel.grid.minor = element_blank(),
               legend.key = element_rect(colour = NA),
               legend.position = "bottom",
               legend.direction = "horizontal",
               legend.key.size= unit(0.2, "cm"),
               legend.margin = margin(t = 0, unit='cm'),
               legend.title = element_text(face="italic"),
               plot.margin=unit(c(10,5,5,5),"mm"),
               strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
               strip.text = element_text()
          ))
      
}

scale_fill_Publication <- function(...){
      library(scales)
      discrete_scale("fill","Publication",manual_pal(values = c("#386cb0","#fdb462","#7fc97f","#ef3b2c","#662506","#a6cee3","#fb9a99","#984ea3","#ffff33")), ...)

}

scale_colour_Publication <- function(...){
      library(scales)
      discrete_scale("colour","Publication",manual_pal(values = c("#386cb0","#fdb462","#7fc97f","#ef3b2c","#662506","#a6cee3","#fb9a99","#984ea3","#ffff33")), ...)

}

## loading packages
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg38.knownGene) # Need to run for the right genome version. 
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
library(org.Hs.eg.db)
library(ggupset)
library(ReactomePA)
library(clusterProfiler) # the GO part
library(grid)
library(gghighlight)
library(DOSE)
library(ggplot2)
library(cowplot)
library(wesanderson)
library(ReactomePA)
library("pathview")
library(annotate)
library(GenomicFeatures)

options(ChIPseeker.downstreamDistance = 3000) # this can be used to defined the options. 
options(ChIPseeker.ignore_1st_exon = TRUE)
options(ChIPseeker.ignore_1st_intron = TRUE)
options(ChIPseeker.ignore_downstream = TRUE) # put all into intergenic regions
options(ChIPseeker.ignore_promoter_subcategory = TRUE)

```

## basic environment setup to run ArchR v1.0.2 or the most recent develop version
Find **ArchR** information at <https://www.archrproject.com/index.html>
Additional information about running **ArchR** multiomic pipeline: 
<https://greenleaflab.github.io/ArchR_2020/Ex-Analyze-Multiome.html>
```{r}
set.seed(1)
addArchRThreads(threads = 100)
addArchRGenome("hg38")
getArchRGenome()
```

## Use The Wilcoxon rank sum test in the ArchR getMarkerFeatures() for differential comparison
For each C9orf72 ALS/FTD sample, pseudo-bulk data from each 35 distinct clusters were compared to that from the healthy control samples using ArchR getMarkerFeatures() with default parameters. The Wilcoxon rank sum test was used since this is most widely acceptable single-cell differential comparison method. The three groups of patients, grouped based on pTDP-43 levels, were compared with control samples. GeneExpressionMatrix and PeakMatrix generated by ArchR were used for differential genes and differential accessibility analysis, respectively. Genes and chromatin accessibility regions were considered significant differentially if they have an FDR-corrected p-value < 0.05 and an absolute log2(fold change) > 0.5 relative to the control group. To ensure a high confidence of differential gene expression comparison, we performed at least 5 permutation tests by randomizing the sample tag per nucleus while keeping the same proportion of each cell type per groups of patients. None of the permutation tests identified any significant differently expressed genes (FDR-corrected p-value < 0.05), ensuring high confidence in the differential gene expression comparison.   

## Differential gene expression
### 1 ### preparing the matrix and comparison

```{r}
## make output directory 
## fGenes = Differential gene expression
fGenes_dir 
  dir.create(fGenes_dir, recursive = TRUE)
fGenes_dirnoflt 
  dir.create(fGenes_dirnoflt, recursive = TRUE)
fGenes_dirFDR 
  dir.create(fGenes_dirFDR, recursive = TRUE)
fGenes_dirFDRMA 
  dir.create(fGenes_dirFDRMA, recursive = TRUE)

# use for loop for the comparison:
for (U in 1:32) {
  print(paste("START with--U --",U, sep=""))
TYPE <- gsub("-", "_",clu38TDP_compo$type[U])

  print(paste("START with U --",U, ":: TYPE is ", TYPE, sep=""))
CTL <- clu38TDP_compo$control[U]
highTDP <- clu38TDP_compo$highTDP[U]
medTDP <- clu38TDP_compo$medTDP[U]
noTDP <- clu38TDP_compo$noTDP[U]
  print(paste("compare::", CTL, "with:", highTDP, medTDP, noTDP, sep=' '))
  
  print(paste(TYPE,"-- STEP 1 -- fGene comp",sep=""))
print("=== v highTDP")
fGenes_vhighTDP <- getMarkerFeatures(
  ArchRProj = ArchR_C9, 
  useMatrix = "GeneExpressionMatrix",
  groupBy = "clu38TDP",testMethod = "wilcoxon",bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = highTDP,bgdGroups = CTL)
print("=== v medTDP")
fGenes_vmedTDP <- getMarkerFeatures(
  ArchRProj = ArchR_C9, 
  useMatrix = "GeneExpressionMatrix",
  groupBy = "clu38TDP",testMethod = "wilcoxon",bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = medTDP,bgdGroups = CTL)
print("=== v noTDP")
fGenes_vnoTDP <- getMarkerFeatures(
  ArchRProj = ArchR_C9, 
  useMatrix = "GeneExpressionMatrix",
  groupBy = "clu38TDP",testMethod = "wilcoxon",bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = noTDP,bgdGroups = CTL)
# save out to a name 
assign(paste("fGenes_",TYPE,"_vhighTDP",sep=""), fGenes_vhighTDP)
assign(paste("fGenes_",TYPE,"_vmedTDP",sep=""), fGenes_vmedTDP)
assign(paste("fGenes_",TYPE,"_vnoTDP",sep=""), fGenes_vnoTDP)

## save RDS
print("== SAVE RDS")
  saveRDS(fGenes_vhighTDP, paste(fGenes_dir,paste("fGenes_",TYPE,"_vhighTDP",sep=""),".rds",sep="")) 
  saveRDS(fGenes_vmedTDP, paste(fGenes_dir,paste("fGenes_",TYPE,"_vmedTDP",sep=""),".rds",sep="")) 
  saveRDS(fGenes_vnoTDP, paste(fGenes_dir,paste("fGenes_",TYPE,"_vnoTDP",sep=""),".rds",sep="")) 

# filter: 
print("== Filter and write out") # the first item on the list 
fGenes_vhighTDP_FDR <- getMarkers(fGenes_vhighTDP, cutOff = "FDR <= 0.05 & abs(Log2FC) > 0.5")
  write.table(fGenes_vhighTDP_FDR[[1]],paste(fGenes_dirFDR,paste("fGenes_",TYPE,"_vhighTDP",sep=""),".FDR.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
fGenes_vmedTDP_FDR <- getMarkers(fGenes_vmedTDP, cutOff = "FDR <= 0.05 & abs(Log2FC) > 0.5")
  write.table(fGenes_vmedTDP_FDR[[1]],paste(fGenes_dirFDR,paste("fGenes_",TYPE,"_vmedTDP",sep=""),".FDR.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
fGenes_vnoTDP_FDR <- getMarkers(fGenes_vnoTDP, cutOff = "FDR <= 0.05 & abs(Log2FC) > 0.5")
  write.table(fGenes_vnoTDP_FDR[[1]],paste(fGenes_dirFDR,paste("fGenes_",TYPE,"_vnoTDP",sep=""),".FDR.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
    # save out to a name 
    assign(paste("fGenes_",TYPE,"_vhighTDP","_FDR",sep=""), fGenes_vhighTDP_FDR)
    assign(paste("fGenes_",TYPE,"_vmedTDP","_FDR",sep=""), fGenes_vmedTDP_FDR)
    assign(paste("fGenes_",TYPE,"_vnoTDP","_FDR",sep=""), fGenes_vnoTDP_FDR)

## get the whole list out 
print("== write out whole list") # the first item on the list 
fGenes_vhighTDP_noflt <- getMarkers(fGenes_vhighTDP, cutOff = "abs(Log2FC)>=0")
  write.table(fGenes_vhighTDP_noflt[[1]],paste(fGenes_dirnoflt,paste("fGenes_",TYPE,"_vhighTDP",sep=""),".noflt.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
fGenes_vmedTDP_noflt <- getMarkers(fGenes_vmedTDP, cutOff = "abs(Log2FC)>=0")
  write.table(fGenes_vmedTDP_noflt[[1]],paste(fGenes_dirnoflt,paste("fGenes_",TYPE,"_vmedTDP",sep=""),".noflt.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
fGenes_vnoTDP_noflt <- getMarkers(fGenes_vnoTDP, cutOff = "abs(Log2FC)>=0")
  write.table(fGenes_vnoTDP_noflt[[1]],paste(fGenes_dirnoflt,paste("fGenes_",TYPE,"_vnoTDP",sep=""),".noflt.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
}

## fGenes post prcoessing 
fGenes_lst <- grep("FDR$", ls(pattern="^fGenes_C"), value = TRUE)
fGenes_names ## the complied list of significant genes
 fGenes_namesGR <- makeGRangesFromDataFrame(fGenes_names, keep.extra.columns=TRUE)

hs <- org.Hs.eg.db

fGenes_names_hs_ALIAS <-
  AnnotationDbi::select(hs, 
       keys = fGenes_names$name,
       columns = c("ENTREZID", "SYMBOL", "GENETYPE", "ALIAS", "ENSEMBL"),
       keytype = "ALIAS")
unique(fGenes_names_hs_ALIAS$GENETYPE)
length(unique(fGenes_names_hs_ALIAS$ALIAS))
fGenes_names_hs_ALIAS[is.na(fGenes_names_hs_ALIAS)] <- "NOinfo"

fGenes_names_hs_ALIAS_NOinfo <- filter(fGenes_names_hs_ALIAS, GENETYPE=="NOinfo")
length(unique(fGenes_names_hs_ALIAS_NOinfo$ALIAS))
length(unique(filter(fGenes_names_hs_ALIAS, GENETYPE!="NOinfo")$ALIAS))

unique(fGenes_names_hs_ALISA_info$GENETYPE)
[1] "other"             "ncRNA"             "protein-coding"    "pseudo"            "unknown"           "biological-region" "tRNA"   

## Clean up the list, only keep the ncRNA and protein-coding genes
  # add count, like sort uniq -c in UNIX
fGenes_names_lst <- filter(fGenes_names_hs_ALIAS, GENETYPE=="protein-coding" | GENETYPE=="ncRNA")  %>% add_count(ALIAS)

## this is the final list 
fGene_final <- rbind(fGenes_names_lst_uniq, fGenes_names_lst_dup_580, fGenes_names_lst_dup_miss10) %>% filter(ENTREZID!=122394733) %>% select(-n)

### Then multiple merging somehow
    fGene_final$name <- fGene_final$ALIAS
head(fGene_final)
    #    ALIAS ENTREZID SYMBOL       GENETYPE   name
    # 1   XIST     7503   XIST          ncRNA   XIST
    # 2 EEF1A1     1915 EEF1A1 protein-coding EEF1A1
    # 3    FTL     2512    FTL protein-coding    FTL
    # 4  GAPDH     2597  GAPDH protein-coding  GAPDH
    # 5 TMSB10     9168 TMSB10 protein-coding TMSB10
    # 6   FTH1     2495   FTH1 protein-coding   FTH1
fGenes_names

## but need to make the list better. 
fGenes_lst <- list(fGenes_lst) # 96 of them 
fGenes_df <- as.data.frame(fGenes_lst)
colnames(fGenes_df) <- c("list")

## out is slimmer matrix 
for (U in 1:96) {
  print(U)
    item <- fGenes_df$list[U]
    ID <- gsub("_FDR", "", gsub("fGenes_", "", item))
    temp1 <- as.data.frame(get(item)[[1]]) %>% select(name,Log2FC)
    colnames(temp1) <- c("name", ID)
  assign(paste(ID,"_c",sep=""), temp1)
}

## then grep the list 
fGenes_clst <- grep("TDP_c$", ls(), value = TRUE)
## clean up the table
fGenes_df$FDRlst <- paste(gsub("_FDR", "", gsub("fGenes_", "", fGenes_df$list)), "_c", sep="")
fGenes_df$types <- sapply(strsplit(fGenes_df$list, "_"), function(x) x[3]) 
fGenes_df$clu38 <- sapply(strsplit(fGenes_df$list, "_"), function(x) x[2]) 
fGenes_df$grp <- sapply(strsplit(fGenes_df$list, "_"), function(x) x[4]) 

head(fGenes_df)

unique(fGenes_df$type)
table(fGenes_df$type)

## then ready to merge 

### matching
fGenes_clean <- Reduce(function(x,y) merge(x = x, y = y, by = "name", all.x = TRUE, all.y = FALSE, incomparables ="NOMATCH"), 
       list(fGene_final, fGenes_names))

################################
## then start making the big combined list, one group at a time
####### 1. MG #######
  for (i in 1:12) {
    dat <- get(filter(fGenes_df, types =="MG")$FDRlst[i])
    MG_lst[[i]] <- dat # add it to your list}

fGenes_mat_MG <- Reduce(function(x,y) merge(x = x, y = y, by = "name", all.x = TRUE, all.y = TRUE), 
       list(fGenes_clean, MG_lst[1], MG_lst[2],MG_lst[3],MG_lst[4],MG_lst[5],MG_lst[6],MG_lst[7],MG_lst[8], MG_lst[9],MG_lst[10],MG_lst[11],MG_lst[12]))

fGenes_mat_MG[is.na(fGenes_mat_MG)] <- "NOMATCH"
head(fGenes_mat_MG)
########  MG END #################
####### 2. ASC #######
ASC_lst <- list()
  for (i in 1:9) {
    dat <- get(filter(fGenes_df, types =="ASC")$FDRlst[i])
    ASC_lst[[i]] <- dat }
fGenes_mat_ASC <- Reduce(function(x,y) merge(x = x, y = y, by = "name", all.x = TRUE, all.y = TRUE), 
       list(fGenes_clean, ASC_lst[1], ASC_lst[2],ASC_lst[3],ASC_lst[4],ASC_lst[5],ASC_lst[6],ASC_lst[7],ASC_lst[8], ASC_lst[9]))

fGenes_mat_ASC[is.na(fGenes_mat_ASC)] <- "NOMATCH"
head(fGenes_mat_ASC)
########  ASC END #################
####### 3. OPC  #######
OPC_lst <- list()
  for (i in 1:9) {
    dat <- get(filter(fGenes_df, types =="OPC")$FDRlst[i])
    OPC_lst[[i]] <- dat }
fGenes_mat_OPC <- Reduce(function(x,y) merge(x = x, y = y, by = "name", all.x = TRUE, all.y = TRUE), 
       list(fGenes_clean, OPC_lst[1], OPC_lst[2],OPC_lst[3],OPC_lst[4],OPC_lst[5],OPC_lst[6],OPC_lst[7],OPC_lst[8], OPC_lst[9]))

fGenes_mat_OPC[is.na(fGenes_mat_OPC)] <- "NOMATCH"
head(fGenes_mat_OPC)
########  OPC END #################
####### 4. ODC  #######
ODC_lst <- list()
  for (i in 1:18) {
    dat <- get(filter(fGenes_df, types =="ODC")$FDRlst[i])
    ODC_lst[[i]] <- dat }
fGenes_mat_ODC <- Reduce(function(x,y) merge(x = x, y = y, by = "name", all.x = TRUE, all.y = TRUE), 
       list(fGenes_clean, ODC_lst[1], ODC_lst[2],ODC_lst[3],ODC_lst[4],ODC_lst[5],ODC_lst[6],ODC_lst[7],ODC_lst[8], ODC_lst[9],ODC_lst[10], ODC_lst[11],ODC_lst[12],ODC_lst[13],ODC_lst[14],ODC_lst[15],ODC_lst[16],ODC_lst[17], ODC_lst[18]))

fGenes_mat_ODC[is.na(fGenes_mat_ODC)] <- "NOMATCH"
head(fGenes_mat_ODC)
########  ODC END #################
####### 5. ENDO  #######
ENDO_lst <- list()
  for (i in 1:3) {
    dat <- get(filter(fGenes_df, types =="ENDO")$FDRlst[i])
    ENDO_lst[[i]] <- dat }
fGenes_mat_ENDO <- Reduce(function(x,y) merge(x = x, y = y, by = "name", all.x = TRUE, all.y = TRUE), 
       list(fGenes_clean, ENDO_lst[1], ENDO_lst[2],ENDO_lst[3]))

fGenes_mat_ENDO[is.na(fGenes_mat_ENDO)] <- "NOMATCH"
head(fGenes_mat_ENDO)
########  ENDO END #################
####### 6. EX  #######
EX_lst <- list()
  for (i in 1:27) {
    dat <- get(filter(fGenes_df, types =="EX")$FDRlst[i])
    EX_lst[[i]] <- dat }
fGenes_mat_EX <- Reduce(function(x,y) merge(x = x, y = y, by = "name", all.x = TRUE, all.y = TRUE), 
       list(fGenes_clean, EX_lst[1], EX_lst[2],EX_lst[3],EX_lst[4],EX_lst[5],EX_lst[6],EX_lst[7],EX_lst[8], EX_lst[9],EX_lst[10], EX_lst[11],EX_lst[12],EX_lst[13],EX_lst[14],EX_lst[15],EX_lst[16],EX_lst[17], EX_lst[18], EX_lst[19], EX_lst[20], EX_lst[21], EX_lst[22],EX_lst[23],EX_lst[24],EX_lst[25],EX_lst[26],EX_lst[27]))

fGenes_mat_EX[is.na(fGenes_mat_EX)] <- "NOMATCH"
head(fGenes_mat_EX)
########  EX END #################
####### 7. IN  #######
IN_lst <- list()
  for (i in 1:18) {
    dat <- get(filter(fGenes_df, types =="IN")$FDRlst[i])
    IN_lst[[i]] <- dat }
fGenes_mat_IN <- Reduce(function(x,y) merge(x = x, y = y, by = "name", all.x = TRUE, all.y = TRUE), 
       list(fGenes_clean, IN_lst[1], IN_lst[2],IN_lst[3],IN_lst[4],IN_lst[5],IN_lst[6],IN_lst[7],IN_lst[8], IN_lst[9],IN_lst[10], IN_lst[11],IN_lst[12],IN_lst[13],IN_lst[14],IN_lst[15],IN_lst[16],IN_lst[17], IN_lst[18]))

fGenes_mat_IN[is.na(fGenes_mat_IN)] <- "NOMATCH"
head(fGenes_mat_IN)
########  IN END #################

### all the gene list 
get(ls(pattern="^fGenes_mat")[1])

  ## export out just in case 
for (i in 1:7) {
write.table(get(ls(pattern="^fGenes_mat")[i]), 
            paste(, ls(pattern="^fGenes_mat")[i], ".cMat.txt", sep=""),
      append= F, sep='\t', 
      row.name = FALSE, col.names = TRUE, quote = FALSE)  
}


####### START combining into a bigger table -fGenes #################
## the start dir: 15 clusters noe count (don't iimport the zero), so total 73 only
dir # where the differential list of genes are
  
fGenesFDR.list <- list.files(dir, full.names = TRUE)

## out is clean, then merge all files, add the column 
for (U in 1:73) {
  print(U)
    item <- fGenesFDR.list.df$names[U]
    ID <- gsub("_FDR", "", gsub("fGenes_", "", item))
    temp1 <- as.data.frame(get(item)) %>% select(-MeanDiff,-strand, -idx)
    temp1$types <- ID
    temp2 <- merge(fGenes_clean,temp1,by = "name", all.x = FALSE, all.y = TRUE)
    temp2[is.na(temp2)] <- "NOTgenes"
    temp2 <- filter(temp2, ALIAS!="NOTgenes") %>% select(-X96, -seqnames.x, -start.x, -end.x)
    colnames(temp2) <- c("name","ALIAS","ENTREZID","SYMBOL","GENETYPE", "seqnames","start","end","Log2FC","FDR","types")
    temp2$celltype <- fGenesFDR.list.df$types[U]
    temp2$clu38 <- fGenesFDR.list.df$clu38[U]
    temp2$grp <- fGenesFDR.list.df$grp[U]
  assign(paste(ID,"_x",sep=""), temp2)
}  

## rbind into one file
  fGenesFDR_alist <- mget(fGenesFDR.list.df$FDRlst)
  fGenesFDR_list_all <- rbindlist(fGenesFDR_alist, fill=TRUE) %>% as.data.frame()

## then count the numbers
  head(fGenesFDR_list_all)
  fGenesFDR_list_all$celltype_grp <- paste(fGenesFDR_list_all$celltype, fGenesFDR_list_all$grp, sep="_")
  fGenesFDR_list_all$UpDown <- ifelse(fGenesFDR_list_all$Log2FC>0, "UpGene", "DownGene") %>% as.character()

    fGenesFDR_list_all$celltype_grp_change <- paste(fGenesFDR_list_all$celltype_grp, fGenesFDR_list_all$UpDown, sep="_")

fGenesFDR_list_all %>% count(fGenesFDR_list_all, "UpDown")
  
fGenesFDR_list_all %>% 
   group_by(UpDown) %>% 
   summarise(freq = n())

fGenesFDR_list_all_celltype_grp <- fGenesFDR_list_all %>% 
   group_by(celltype_grp) %>% 
   summarise(freq = n()) %>% as.data.frame()

fGenesFDR_list_all %>%
  group_by(celltype_grp_change) %>%
  summarise(gene_name_ct = n_distinct(name))  

fGenesFDR_list_all_UpGene_ct_byCelltypeFreq <- 
    fGenesFDR_list_all %>% select(name, celltype, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_UpGene_ct_byCelltypeFreq$type <- "UpGene"
  fGenesFDR_list_all_DownGene_ct_byCelltypeFreq <- 
    fGenesFDR_list_all %>% select(name, celltype, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_DownGene_ct_byCelltypeFreq$type <- "DownGene"

  
 fGenesFDR_list_all_UpGene_ct_byCelltypeFreq_mg <-  
   merge(fGenesFDR_list_all_UpGene_ct_byCelltypeFreq, fGenesFDR_list_all_DownGene_ct_byCelltypeFreq, by="name_ct", suffixes = c("_UpGene","_DownGene"))
  fGenesFDR_list_all_UpGene_ct_byCelltypeFreq_long <- rbind(fGenesFDR_list_all_UpGene_ct_byCelltypeFreq, fGenesFDR_list_all_DownGene_ct_byCelltypeFreq)

pdf(file="fGenesFDR_list_all_UpGene_ct_byCelltypeFreq_mg_plot.pdf", 
     width = 2.5, height = 1.8)
  fGenesFDR_list_all_UpGene_ct_byCelltypeFreq_long %>% ggplot(aes(x=factor(name_ct), y=freq, fill=type)) +
    geom_bar(stat="identity", color="black", position=position_dodge(), width = 0.85, size=0.2)+ylab("Number of DEG") + xlab("Number of perturbed cell-types") +
  scale_fill_manual(values=c('#15677A','#D89000')) + theme_Publication()

dev.off()   

## then do the same GO in the super shared categories

fGenesFDR_list_all$celltype %>% unique()
  # [1] "ODC"  "EX"   "IN"   "ASC"  "MG"   "OPC"  "ENDO"
fGenesFDR_list_all$major %>% unique()
  fGenesFDR_list_all$major <- fGenesFDR_list_all$celltype
  fGenesFDR_list_all$major <- gsub("IN", "Neu", gsub("EX", "Neu", fGenesFDR_list_all$major))
  fGenesFDR_list_all$major <- gsub("ASC", "nonNeu", gsub("ODC", "nonNeu", gsub("MG", "nonNeu", gsub("OPC", "nonNeu", gsub("ENDO", "nonNeu", fGenesFDR_list_all$major)))))

  fGenesFDR_list_all_UpGene_ct_bymajorFreq <- 
    fGenesFDR_list_all %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_UpGene_ct_bymajorFreq$type <- "UpGene"
    
  fGenesFDR_list_all_DownGene_ct_bymajorFreq <- 
    fGenesFDR_list_all %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_DownGene_ct_bymajorFreq$type <- "DownGene"
  
fGenesFDR_list_all_ct_bymajorFreq <- rbind(fGenesFDR_list_all_UpGene_ct_bymajorFreq, fGenesFDR_list_all_DownGene_ct_bymajorFreq)

# 1 = unique to Neurons or nonNeurons, 2 = common among the Neurons & nonNeurons. 
pdf(file="fGenesFDR_list_all_ct_bymajorFreq_plot.pdf", 
     width = 1.8, height = 1.8)
  fGenesFDR_list_all_ct_bymajorFreq %>% ggplot(aes(x=factor(name_ct), y=freq, fill=type)) +
    geom_bar(stat="identity", color="black", position=position_dodge(), width = 0.85, size=0.2)+ylab("Number of DEG") + xlab("Number of perturbed cell-types") +
  scale_fill_manual(values=c('#15677A','#D89000')) + theme_Publication()

dev.off()       
    
## then how to split Neu or nonNeu specific bars 
    fGenesFDR_list_all_UpGene_ct_Neu <- 
      fGenesFDR_list_all %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% filter(major=="Neu") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 6400 (- 2937 = 3464)
    fGenesFDR_list_all_UpGene_ct_Neu$type <- "UpGene_Neu"
    fGenesFDR_list_all_UpGene_ct_Neu$freq=fGenesFDR_list_all_UpGene_ct_Neu$freq-2937
    
    fGenesFDR_list_all_UpGene_ct_nonNeu <- 
    fGenesFDR_list_all %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% filter(major=="nonNeu") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 4949 (- 2937 = 2012)
    fGenesFDR_list_all_UpGene_ct_nonNeu$type <- "UpGene_nonNeu"
    fGenesFDR_list_all_UpGene_ct_nonNeu$freq=fGenesFDR_list_all_UpGene_ct_nonNeu$freq-2937
    
    fGenesFDR_list_all_DownGene_ct_Neu <- 
     fGenesFDR_list_all %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% filter(major=="Neu") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 9687 (- 6172 = 3515)
      fGenesFDR_list_all_DownGene_ct_Neu$type <- "DownGene_Neu"
      fGenesFDR_list_all_DownGene_ct_Neu$freq=fGenesFDR_list_all_DownGene_ct_Neu$freq-6172
    fGenesFDR_list_all_DownGene_ct_nonNeu <- 
      fGenesFDR_list_all %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% filter(major=="nonNeu") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 8224 (- 6172 = 2052)
      fGenesFDR_list_all_DownGene_ct_nonNeu$type <- "DownGene_nonNeu"
      fGenesFDR_list_all_DownGene_ct_nonNeu$freq=fGenesFDR_list_all_DownGene_ct_nonNeu$freq-6172
 
fGenesFDR_list_all_ct_bymajorFreq2 <- 
  rbind(fGenesFDR_list_all_UpGene_ct_bymajorFreq, fGenesFDR_list_all_DownGene_ct_bymajorFreq) %>% filter(name_ct!=1)
fGenesFDR_list_all_ct_bymajorFreq2$UpDown <- fGenesFDR_list_all_ct_bymajorFreq2$type
fGenesFDR_list_all_ct_bymajorFreq2$type2 <- "common"
fGenesFDR_list_all_ct_bymajorFreq3 <- 
  rbind(fGenesFDR_list_all_UpGene_ct_Neu, fGenesFDR_list_all_UpGene_ct_nonNeu, fGenesFDR_list_all_DownGene_ct_Neu, fGenesFDR_list_all_DownGene_ct_nonNeu)    
fGenesFDR_list_all_ct_bymajorFreq3$UpDown <- sapply(strsplit(fGenesFDR_list_all_ct_bymajorFreq3$type , "_"), function(x) x[1])
fGenesFDR_list_all_ct_bymajorFreq3$type2 <- sapply(strsplit(fGenesFDR_list_all_ct_bymajorFreq3$type , "_"), function(x) x[2])

fGenesFDR_list_all_ct_bymajorFreq_split <-  
rbind(fGenesFDR_list_all_ct_bymajorFreq2 %>% select(freq,UpDown, type2), fGenesFDR_list_all_ct_bymajorFreq3%>%select(freq,UpDown, type2))

library("wesanderson")

## re_plot
pdf(file="fGenesFDR_list_all_ct_bymajorFreq_split_plot.pdf", 
     width = 1.8, height = 1.8)
  fGenesFDR_list_all_ct_bymajorFreq_split %>% mutate(UpDown = factor(UpDown, levels=c("UpGene", "DownGene"))) %>% ggplot(aes(x=factor(UpDown), y=freq, fill=type2)) +
    geom_bar(stat="identity", color="black", position=position_dodge(), width = 0.85, size=0.2)+ylab("Number of DEG") + xlab("Number of perturbed cell-types") +
  scale_fill_manual(values=wes_palette(n = 3, "GrandBudapest1")) + theme_Publication()
dev.off()  
    
```

### 2 gene ontology analysis of the differential genes
```{r}
library("stringr")
## the up and down commons: 
  fGenesFDR_list_all_UpGene_common <- 
    fGenesFDR_list_all %>% select(ENTREZID, major, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% group_by(ENTREZID) %>% summarise(name_ct = n()) %>% filter(name_ct==2) %>% as.data.frame() 
fGenesFDR_list_all_UpGene_common$name_ct <- gsub(2, "Up_common", fGenesFDR_list_all_UpGene_common$name_ct)

  fGenesFDR_list_all_DownGene_common <- 
    fGenesFDR_list_all %>% select(ENTREZID, major, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% group_by(ENTREZID) %>% summarise(name_ct = n()) %>% filter(name_ct==2) %>% as.data.frame() 
fGenesFDR_list_all_DownGene_common$name_ct <- gsub(2, "Up_common", fGenesFDR_list_all_DownGene_common$name_ct)
    
## GO on the common 
fGene_common_lst <- list(Common_Up=fGenesFDR_list_all_UpGene_common, Common_Down=fGenesFDR_list_all_DownGene_common)
fGene_common_lst_genes = lapply(fGene_common_lst, function(i) as.data.frame(i)$ENTREZID)
  ## KEGG
fGene_common_lst_genes_KEGG <- compareCluster(geneCluster   = fGene_common_lst_genes,
                         fun           = "enrichKEGG",
                         pvalueCutoff  = 0.05, #keyType = "ENTREZID",
                         pAdjustMethod = "BH")
fGene_common_lst_genes_goBP <- compareCluster(geneCluster   = fGene_common_lst_genes, OrgDb=org.Hs.eg.db,ont = "BP",
                         fun           = "enrichGO",
                         pvalueCutoff  = 0.05, #keyType = "ENTREZID",
                         pAdjustMethod = "BH")
fGene_common_lst_genes_goMF <- compareCluster(geneCluster   = fGene_common_lst_genes, OrgDb=org.Hs.eg.db,ont = "MF",
                         fun           = "enrichGO",
                         pvalueCutoff  = 0.05, #keyType = "ENTREZID",
                         pAdjustMethod = "BH")

pdf(file="fGene_common_lst_genes_GOs_plot.pdf", 
     width = 10, height = 10, onefile=TRUE)
dotplot(fGene_common_lst_genes_KEGG, showCategory = 50, title = "KEGG Pathway Enrichment Analysis") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(fGene_common_lst_genes_goBP, showCategory = 50, title = "GO-biological processes Enrichment") + scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(fGene_common_lst_genes_goMF, showCategory = 50, title = "GO-Molecular function Enrichment") + scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goBP, level=4), showCategory = 50, title = "goBP_slim level 4") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))s
dotplot(gofilter(fGene_common_lst_genes_goBP, level=3), showCategory = 50, title = "goBP_slim level 3") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goBP, level=2), showCategory = 50, title = "goBP_slim level 2") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goMF, level=4), showCategory = 50, title = "goMF_slim level 4") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goMF, level=3), showCategory = 50, title = "goMF_slim level 3") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goMF, level=2), showCategory = 50, title = "goMF_slim level 2") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dev.off()

pdf(file="fGene_common_lst_genes_GOs_plot.pdf", 
     width = 10, height = 10, onefile=TRUE)
dotplot(fGene_common_lst_genes_KEGG, showCategory = 50, title = "KEGG Pathway Enrichment Analysis") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(fGene_common_lst_genes_goBP, showCategory = 50, title = "GO-biological processes Enrichment") + scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(fGene_common_lst_genes_goMF, showCategory = 50, title = "GO-Molecular function Enrichment") + scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goBP, level=4), showCategory = 50, title = "goBP_slim level 4") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goBP, level=3), showCategory = 50, title = "goBP_slim level 3") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goBP, level=2), showCategory = 50, title = "goBP_slim level 2") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goMF, level=4), showCategory = 50, title = "goMF_slim level 4") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goMF, level=3), showCategory = 50, title = "goMF_slim level 3") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dotplot(gofilter(fGene_common_lst_genes_goMF, level=2), showCategory = 50, title = "goMF_slim level 2") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dev.off()

### goMF_slim level 3 is the most clear, then do bar plot 
pdf(file="fGene_common_lst_genes_GOs_plot.pdf", 
     width = 10, height = 10, onefile=TRUE)
barplot(gofilter(fGene_common_lst_genes_goMF, level=3),x = "p.adjust", showCategory = 50, title = "goMF_slim level 3", split="ONTOLOGY") +scale_y_discrete(labels=function(x) str_wrap(x, width=100))
dev.off()

library(forcats)
library(ggh4x)
  #devtools::install_github("teunbrand/ggh4x")

## re_plot
pdf(file="fGene_common_lst_genes_goMF_level3.pdf", 
     width = 3.5, height = 2.3)
gofilter(fGene_common_lst_genes_goMF, level=3) %>% as.data.frame() %>% dplyr::select(Cluster, Description, qvalue) %>% mutate(Cluster = factor(Cluster, levels=c("Common_Up", "Common_Down"))) %>%mutate(Description = fct_reorder(Description, -log10(qvalue))) %>% 
  ggplot(aes(x=-log10(qvalue), y=Description, fill=Cluster)) + facet_grid(rows = vars(Cluster), scales = "free") + 
  geom_bar(stat="identity", color="black", position=position_dodge(), width = 0.85, size=0.2)+
  ylab("") + xlab(expression(-log[10]("FDR"))) + force_panelsizes(rows = c(2, 10))+ geom_vline(xintercept = 2, colour="#ED1E79", linetype="dashed")+
  scale_fill_manual(values=c('#D89000', '#15677A')) + theme_Publication()
dev.off()  
    
## what are the genes 
  write.table(setReadable(gofilter(fGene_common_lst_genes_goMF, level=3), 'org.Hs.eg.db', 'ENTREZID') %>% as.data.frame(),
              "fGene_common_lst_genes_goMF_level3_df", 
              append= F, sep='\t', 
              row.name = FALSE, col.names = TRUE, quote = FALSE)   
```

### do it per TDP group 
```{r}
fGenesFDR_list_all <- read.csv2("fGenes_all_comp_list73.csv", sep=',') %>% as.data.frame()


fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>%
   group_by(UpDown) %>% 
   summarise(freq = n())

fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>%
   group_by(UpDown) %>% 
   summarise(freq = n())

fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>%
   group_by(UpDown) %>% 
   summarise(freq = n())

fGenesFDR_list_all_UpDown_grp  <- fGenesFDR_list_all %>% #filter(grp=="vnoTDP") %>%
   group_by(grp,UpDown) %>% 
   summarise(freq = n()) %>% as.data.frame()

## remake for each type -- so it might be more specific?? -- treat each group as seperate Identifies
  # this didnt' work. 
    fGenesFDR_list_all %>%
  group_by(celltype_grp_change) %>% summarise(gene_name_ct = n_distinct(name))  

### how to group by and count (total is 22526, uniq is 15051)
  fGenesFDR_list_all_UpGene_vhighTDP_ct_byCelltypeFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>% select(name, celltype, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_UpGene_vhighTDP_ct_byCelltypeFreq$type <- "UpGene"
  fGenesFDR_list_all_DownGene_vhighTDP_ct_byCelltypeFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>% select(name, celltype, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_DownGene_vhighTDP_ct_byCelltypeFreq$type <- "DownGene"
 fGenesFDR_list_all_UpGene_vhighTDP_ct_byCelltypeFreq_mg <-  
   merge(fGenesFDR_list_all_UpGene_vhighTDP_ct_byCelltypeFreq, fGenesFDR_list_all_DownGene_vhighTDP_ct_byCelltypeFreq, by="name_ct", suffixes = c("_UpGene","_DownGene"))
  fGenesFDR_list_all_UpGene_vhighTDP_ct_byCelltypeFreq_long <- rbind(fGenesFDR_list_all_UpGene_vhighTDP_ct_byCelltypeFreq, fGenesFDR_list_all_DownGene_vhighTDP_ct_byCelltypeFreq)
fGenesFDR_list_all_UpGene_vhighTDP_ct_byCelltypeFreq_long$TDPgrp <- "vhighTDP"

# medTDP
  fGenesFDR_list_all_UpGene_vmedTDP_ct_byCelltypeFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>% select(name, celltype, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_UpGene_vmedTDP_ct_byCelltypeFreq$type <- "UpGene"
  fGenesFDR_list_all_DownGene_vmedTDP_ct_byCelltypeFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>% select(name, celltype, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_DownGene_vmedTDP_ct_byCelltypeFreq$type <- "DownGene"
 fGenesFDR_list_all_UpGene_vmedTDP_ct_byCelltypeFreq_mg <-  
   merge(fGenesFDR_list_all_UpGene_vmedTDP_ct_byCelltypeFreq, fGenesFDR_list_all_DownGene_vmedTDP_ct_byCelltypeFreq, by="name_ct", suffixes = c("_UpGene","_DownGene"))
  fGenesFDR_list_all_UpGene_vmedTDP_ct_byCelltypeFreq_long <- rbind(fGenesFDR_list_all_UpGene_vmedTDP_ct_byCelltypeFreq, fGenesFDR_list_all_DownGene_vmedTDP_ct_byCelltypeFreq)
fGenesFDR_list_all_UpGene_vmedTDP_ct_byCelltypeFreq_long$TDPgrp <- "vmedTDP"

# noTDP
  fGenesFDR_list_all_UpGene_vnoTDP_ct_byCelltypeFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>% select(name, celltype, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_UpGene_vnoTDP_ct_byCelltypeFreq$type <- "UpGene"
  fGenesFDR_list_all_DownGene_vnoTDP_ct_byCelltypeFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>% select(name, celltype, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_DownGene_vnoTDP_ct_byCelltypeFreq$type <- "DownGene"
 fGenesFDR_list_all_UpGene_vnoTDP_ct_byCelltypeFreq_mg <-  
   merge(fGenesFDR_list_all_UpGene_vnoTDP_ct_byCelltypeFreq, fGenesFDR_list_all_DownGene_vnoTDP_ct_byCelltypeFreq, by="name_ct", suffixes = c("_UpGene","_DownGene"))
  fGenesFDR_list_all_UpGene_vnoTDP_ct_byCelltypeFreq_long <- rbind(fGenesFDR_list_all_UpGene_vnoTDP_ct_byCelltypeFreq, fGenesFDR_list_all_DownGene_vnoTDP_ct_byCelltypeFreq)
fGenesFDR_list_all_UpGene_vnoTDP_ct_byCelltypeFreq_long$TDPgrp <- "vnoTDP"

## bind everyone, then facet

fGenesFDR_list_all_UpGene_TDPgrp_ct_byCelltypeFreq_long  <- rbind(fGenesFDR_list_all_UpGene_vhighTDP_ct_byCelltypeFreq_long, 
                                                                  fGenesFDR_list_all_UpGene_vmedTDP_ct_byCelltypeFreq_long,
                                                                  fGenesFDR_list_all_UpGene_vnoTDP_ct_byCelltypeFreq_long)
pdf(file="fGenesFDR_list_all_UpGene_TDPgrp_ct_byCelltypeFreq_mg_plot.pdf", 
     width = 2.5, height = 2.8, onefile=TRUE)
  fGenesFDR_list_all_UpGene_TDPgrp_ct_byCelltypeFreq_long %>% ggplot(aes(x=factor(name_ct), y=freq, fill=type)) + 
    geom_bar(stat="identity", color="black", position=position_dodge(), width = 0.85, size=0.2)+ylab("Number of DEG") + xlab("Number of perturbed cell-types") +
    #scale_y_break(c(1500, 2500)) + scale_y_log10()+
    facet_grid(TDPgrp ~ .) + scale_fill_manual(values=c('#b35806','#542788')) + theme_Publication()
  fGenesFDR_list_all_UpGene_TDPgrp_ct_byCelltypeFreq_long %>% filter(name_ct !=1)%>% ggplot(aes(x=factor(name_ct), y=freq, fill=type)) +
    geom_bar(stat="identity", color="black", position=position_dodge(), width = 0.85, size=0.2)+ylab("Number of DEG") + xlab("Number of perturbed cell-types") +
    facet_grid(TDPgrp ~ .) + scale_fill_manual(values=c('#b35806','#542788')) + theme_Publication()
dev.off()       
```

## do each group, but neu vs nonNeu
```{r}
##  do the same GO in the neu and nonNeu shared category
  fGenesFDR_list_all_UpGene_vhighTDP_ct_bymajorFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% 
    group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_UpGene_vhighTDP_ct_bymajorFreq$type <- "UpGene"
    fGenesFDR_list_all_UpGene_vhighTDP_ct_bymajorFreq$TDPgrp <- "vhighTDP"
    
  fGenesFDR_list_all_DownGene_vhighTDP_ct_bymajorFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% 
    group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_DownGene_vhighTDP_ct_bymajorFreq$type <- "DownGene"
    fGenesFDR_list_all_DownGene_vhighTDP_ct_bymajorFreq$TDPgrp <- "vhighTDP"
        fGenesFDR_list_vhighTDP_ct_bymajorFreq <- rbind(fGenesFDR_list_all_UpGene_vhighTDP_ct_bymajorFreq, fGenesFDR_list_all_DownGene_vhighTDP_ct_bymajorFreq)
# vmedTDP        
  fGenesFDR_list_all_UpGene_vmedTDP_ct_bymajorFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% 
    group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_UpGene_vmedTDP_ct_bymajorFreq$type <- "UpGene"
    fGenesFDR_list_all_UpGene_vmedTDP_ct_bymajorFreq$TDPgrp <- "vmedTDP"
    
  fGenesFDR_list_all_DownGene_vmedTDP_ct_bymajorFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% 
    group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_DownGene_vmedTDP_ct_bymajorFreq$type <- "DownGene"
    fGenesFDR_list_all_DownGene_vmedTDP_ct_bymajorFreq$TDPgrp <- "vmedTDP"
        fGenesFDR_list_vmedTDP_ct_bymajorFreq <- rbind(fGenesFDR_list_all_UpGene_vmedTDP_ct_bymajorFreq, fGenesFDR_list_all_DownGene_vmedTDP_ct_bymajorFreq)
# vnoTDP        
  fGenesFDR_list_all_UpGene_vnoTDP_ct_bymajorFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% 
    group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_UpGene_vnoTDP_ct_bymajorFreq$type <- "UpGene"
    fGenesFDR_list_all_UpGene_vnoTDP_ct_bymajorFreq$TDPgrp <- "vnoTDP"
    
  fGenesFDR_list_all_DownGene_vnoTDP_ct_bymajorFreq <- 
    fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% 
    group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame()
    fGenesFDR_list_all_DownGene_vnoTDP_ct_bymajorFreq$type <- "DownGene"
    fGenesFDR_list_all_DownGene_vnoTDP_ct_bymajorFreq$TDPgrp <- "vnoTDP"
        fGenesFDR_list_vnoTDP_ct_bymajorFreq <- rbind(fGenesFDR_list_all_UpGene_vnoTDP_ct_bymajorFreq, fGenesFDR_list_all_DownGene_vnoTDP_ct_bymajorFreq)

## bind All : 
    fGenesFDR_list_all_TDPgrp_ct_bymajorFreq <- rbind(fGenesFDR_list_vmedTDP_ct_bymajorFreq, fGenesFDR_list_vnoTDP_ct_bymajorFreq, fGenesFDR_list_vhighTDP_ct_bymajorFreq)
    write.table(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq,
              "fGenes_fGenesFDR_list_all_TDPgrp_ct_bymajorFreq.txt", 
              append= F, sep='\t', 
              row.name = FALSE, col.names = TRUE, quote = FALSE)        

##################     
## then how to split Neu or nonNeu specific bars
#### vhighTDP ####
  fGenesFDR_list_all_UpGene_vhighTDP_ct_Neu <- 
      fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% 
      filter(major=="Neu") %>% unique() %>% group_by(name) %>% 
      summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 6400 (- 2937 = 3464)
    fGenesFDR_list_all_UpGene_vhighTDP_ct_Neu$type <- "UpGene_Neu"
    fGenesFDR_list_all_UpGene_vhighTDP_ct_Neu$TDPgrp <- "vhighTDP"
    fGenesFDR_list_all_UpGene_vhighTDP_ct_Neu$freq=
      fGenesFDR_list_all_UpGene_vhighTDP_ct_Neu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="UpGene" & TDPgrp=="vhighTDP" & name_ct==2)$freq
  fGenesFDR_list_all_UpGene_vhighTDP_ct_nonNeu <- 
    fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% 
      filter(major=="nonNeu") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 4949 (- 2937 = 2012)
    fGenesFDR_list_all_UpGene_vhighTDP_ct_nonNeu$type <- "UpGene_nonNeu"
    fGenesFDR_list_all_UpGene_vhighTDP_ct_nonNeu$TDPgrp <- "vhighTDP"
    fGenesFDR_list_all_UpGene_vhighTDP_ct_nonNeu$freq=
      fGenesFDR_list_all_UpGene_vhighTDP_ct_nonNeu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="UpGene" & TDPgrp=="vhighTDP" & name_ct==2)$freq
  fGenesFDR_list_all_DownGene_vhighTDP_ct_Neu <- 
      fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% 
      filter(major=="Neu") %>% unique() %>% group_by(name) %>% 
      summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 6400 (- 2937 = 3464)
    fGenesFDR_list_all_DownGene_vhighTDP_ct_Neu$type <- "DownGene_Neu"
    fGenesFDR_list_all_DownGene_vhighTDP_ct_Neu$TDPgrp <- "vhighTDP"
    fGenesFDR_list_all_DownGene_vhighTDP_ct_Neu$freq=
      fGenesFDR_list_all_DownGene_vhighTDP_ct_Neu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="DownGene" & TDPgrp=="vhighTDP" & name_ct==2)$freq
  fGenesFDR_list_all_DownGene_vhighTDP_ct_nonNeu <- 
    fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% 
      filter(major=="nonNeu") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 4949 (- 2937 = 2012)
    fGenesFDR_list_all_DownGene_vhighTDP_ct_nonNeu$type <- "DownGene_nonNeu"
    fGenesFDR_list_all_DownGene_vhighTDP_ct_nonNeu$TDPgrp <- "vhighTDP"
    fGenesFDR_list_all_DownGene_vhighTDP_ct_nonNeu$freq=
      fGenesFDR_list_all_DownGene_vhighTDP_ct_nonNeu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="DownGene" & TDPgrp=="vhighTDP" & name_ct==2)$freq    
########
#### vmedTDP ####
  fGenesFDR_list_all_UpGene_vmedTDP_ct_Neu <- 
      fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% 
      filter(major=="Neu") %>% unique() %>% group_by(name) %>% 
      summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 6400 (- 2937 = 3464)
    fGenesFDR_list_all_UpGene_vmedTDP_ct_Neu$type <- "UpGene_Neu"
    fGenesFDR_list_all_UpGene_vmedTDP_ct_Neu$TDPgrp <- "vmedTDP"
    fGenesFDR_list_all_UpGene_vmedTDP_ct_Neu$freq=
      fGenesFDR_list_all_UpGene_vmedTDP_ct_Neu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="UpGene" & TDPgrp=="vmedTDP" & name_ct==2)$freq
  fGenesFDR_list_all_UpGene_vmedTDP_ct_nonNeu <- 
    fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% 
      filter(major=="nonNeu") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 4949 (- 2937 = 2012)
    fGenesFDR_list_all_UpGene_vmedTDP_ct_nonNeu$type <- "UpGene_nonNeu"
    fGenesFDR_list_all_UpGene_vmedTDP_ct_nonNeu$TDPgrp <- "vmedTDP"
    fGenesFDR_list_all_UpGene_vmedTDP_ct_nonNeu$freq=
      fGenesFDR_list_all_UpGene_vmedTDP_ct_nonNeu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="UpGene" & TDPgrp=="vmedTDP" & name_ct==2)$freq
  fGenesFDR_list_all_DownGene_vmedTDP_ct_Neu <- 
      fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% 
      filter(major=="Neu") %>% unique() %>% group_by(name) %>% 
      summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 6400 (- 2937 = 3464)
    fGenesFDR_list_all_DownGene_vmedTDP_ct_Neu$type <- "DownGene_Neu"
    fGenesFDR_list_all_DownGene_vmedTDP_ct_Neu$TDPgrp <- "vmedTDP"
    fGenesFDR_list_all_DownGene_vmedTDP_ct_Neu$freq=
      fGenesFDR_list_all_DownGene_vmedTDP_ct_Neu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="DownGene" & TDPgrp=="vmedTDP" & name_ct==2)$freq
  fGenesFDR_list_all_DownGene_vmedTDP_ct_nonNeu <- 
    fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% 
      filter(major=="nonNeu") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 4949 (- 2937 = 2012)
    fGenesFDR_list_all_DownGene_vmedTDP_ct_nonNeu$type <- "DownGene_nonNeu"
    fGenesFDR_list_all_DownGene_vmedTDP_ct_nonNeu$TDPgrp <- "vmedTDP"
    fGenesFDR_list_all_DownGene_vmedTDP_ct_nonNeu$freq=
      fGenesFDR_list_all_DownGene_vmedTDP_ct_nonNeu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="DownGene" & TDPgrp=="vmedTDP" & name_ct==2)$freq    
########
#### vnoTDP ####
  fGenesFDR_list_all_UpGene_vnoTDP_ct_Neu <- 
      fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% 
      filter(major=="Neu") %>% unique() %>% group_by(name) %>% 
      summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 6400 (- 2937 = 3464)
    fGenesFDR_list_all_UpGene_vnoTDP_ct_Neu$type <- "UpGene_Neu"
    fGenesFDR_list_all_UpGene_vnoTDP_ct_Neu$TDPgrp <- "vnoTDP"
    fGenesFDR_list_all_UpGene_vnoTDP_ct_Neu$freq=
      fGenesFDR_list_all_UpGene_vnoTDP_ct_Neu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="UpGene" & TDPgrp=="vnoTDP" & name_ct==2)$freq
  fGenesFDR_list_all_UpGene_vnoTDP_ct_nonNeu <- 
    fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="UpGene") %>% 
      filter(major=="nonNeu") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 4949 (- 2937 = 2012)
    fGenesFDR_list_all_UpGene_vnoTDP_ct_nonNeu$type <- "UpGene_nonNeu"
    fGenesFDR_list_all_UpGene_vnoTDP_ct_nonNeu$TDPgrp <- "vnoTDP"
    fGenesFDR_list_all_UpGene_vnoTDP_ct_nonNeu$freq=
      fGenesFDR_list_all_UpGene_vnoTDP_ct_nonNeu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="UpGene" & TDPgrp=="vnoTDP" & name_ct==2)$freq
  fGenesFDR_list_all_DownGene_vnoTDP_ct_Neu <- 
      fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% 
      filter(major=="Neu") %>% unique() %>% group_by(name) %>% 
      summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 6400 (- 2937 = 3464)
    fGenesFDR_list_all_DownGene_vnoTDP_ct_Neu$type <- "DownGene_Neu"
    fGenesFDR_list_all_DownGene_vnoTDP_ct_Neu$TDPgrp <- "vnoTDP"
    fGenesFDR_list_all_DownGene_vnoTDP_ct_Neu$freq=
      fGenesFDR_list_all_DownGene_vnoTDP_ct_Neu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="DownGene" & TDPgrp=="vnoTDP" & name_ct==2)$freq
  fGenesFDR_list_all_DownGene_vnoTDP_ct_nonNeu <- 
    fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>% select(name, major, UpDown) %>% filter(UpDown=="DownGene") %>% 
      filter(major=="nonNeu") %>% unique() %>% group_by(name) %>% summarise(name_ct = n()) %>% group_by(name_ct) %>% summarise(freq = n()) %>% as.data.frame() # 4949 (- 2937 = 2012)
    fGenesFDR_list_all_DownGene_vnoTDP_ct_nonNeu$type <- "DownGene_nonNeu"
    fGenesFDR_list_all_DownGene_vnoTDP_ct_nonNeu$TDPgrp <- "vnoTDP"
    fGenesFDR_list_all_DownGene_vnoTDP_ct_nonNeu$freq=
      fGenesFDR_list_all_DownGene_vnoTDP_ct_nonNeu$freq-filter(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq, type=="DownGene" & TDPgrp=="vnoTDP" & name_ct==2)$freq    
########
fGenesFDR_list_all_TDPgrp_ct_bymajorFreq2 <- 
      rbind(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq %>% filter(name_ct==2),
            fGenesFDR_list_all_UpGene_vhighTDP_ct_Neu,fGenesFDR_list_all_DownGene_vhighTDP_ct_Neu,
            fGenesFDR_list_all_UpGene_vhighTDP_ct_nonNeu,fGenesFDR_list_all_DownGene_vhighTDP_ct_nonNeu,
            fGenesFDR_list_all_UpGene_vmedTDP_ct_Neu,fGenesFDR_list_all_DownGene_vmedTDP_ct_Neu,
            fGenesFDR_list_all_UpGene_vmedTDP_ct_nonNeu,fGenesFDR_list_all_DownGene_vmedTDP_ct_nonNeu,
            fGenesFDR_list_all_UpGene_vnoTDP_ct_Neu,fGenesFDR_list_all_DownGene_vnoTDP_ct_Neu,
            fGenesFDR_list_all_UpGene_vnoTDP_ct_nonNeu,fGenesFDR_list_all_DownGene_vnoTDP_ct_nonNeu
            ) 

fGenesFDR_list_all_ct_bymajorFreq2 <- 
  rbind(fGenesFDR_list_all_UpGene_ct_bymajorFreq, fGenesFDR_list_all_DownGene_ct_bymajorFreq) %>% filter(name_ct!=1)

fGenesFDR_list_all_TDPgrp_ct_bymajorFreq2$UpDown <- sapply(strsplit(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq2$type , "_"), function(x) x[1])

fGenesFDR_list_all_TDPgrp_ct_bymajorFreq_com <- fGenesFDR_list_all_TDPgrp_ct_bymajorFreq2 %>% filter(name_ct==2)
fGenesFDR_list_all_TDPgrp_ct_bymajorFreq_com$type2 <- "common"

fGenesFDR_list_all_TDPgrp_ct_bymajorFreq_uniq <- fGenesFDR_list_all_TDPgrp_ct_bymajorFreq2 %>% filter(name_ct==1)
fGenesFDR_list_all_TDPgrp_ct_bymajorFreq_uniq$type2 <- sapply(strsplit(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq_uniq$type , "_"), function(x) x[2])

fGenesFDR_list_all_TDPgrp_ct_bymajorFreq3 <- 
  rbind(fGenesFDR_list_all_TDPgrp_ct_bymajorFreq_com,fGenesFDR_list_all_TDPgrp_ct_bymajorFreq_uniq) %>%select(freq,UpDown, type2, TDPgrp)

library("wesanderson")

## re_plot
pdf(file="fGenesFDR_list_all_TDPgrp_ct_bymajorFreq_split_plot.pdf", 
     width = 1.8, height = 2.5)
  fGenesFDR_list_all_TDPgrp_ct_bymajorFreq3 %>% mutate(UpDown = factor(UpDown, levels=c("UpGene", "DownGene"))) %>% ggplot(aes(x=factor(UpDown), y=freq, fill=type2)) +
    geom_bar(stat="identity", color="black", position=position_dodge(), width = 0.85, size=0.2)+ylab("Number of DEG") + xlab("Number of perturbed cell-types") +
    facet_grid(TDPgrp ~ .)+
  scale_fill_manual(values=wes_palette(n = 3, "GrandBudapest1")) + theme_Publication()
dev.off() 
```


## then GO on the ones that are shared -- per TDP group
```{r}
## rewrite
##vhighTDP
fGenesFDR_list_all_UpGene_vhighTDP_common <- 
  fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>% select(ENTREZID, major, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% 
  group_by(ENTREZID) %>% summarise(name_ct = n()) %>% filter(name_ct==2) %>% as.data.frame() #%>% nrow()
  fGenesFDR_list_all_UpGene_vhighTDP_common$name_ct <- gsub(2, "Up_vhighTDP_common", fGenesFDR_list_all_UpGene_vhighTDP_common$name_ct)
fGenesFDR_list_all_DownGene_vhighTDP_common <- 
  fGenesFDR_list_all %>% filter(grp=="vhighTDP") %>% select(ENTREZID, major, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% 
  group_by(ENTREZID) %>% summarise(name_ct = n()) %>% filter(name_ct==2) %>% as.data.frame() #%>% nrow()
  fGenesFDR_list_all_DownGene_vhighTDP_common$name_ct <- gsub(2, "Up_vhighTDP_common", fGenesFDR_list_all_DownGene_vhighTDP_common$name_ct)
##vmedTDP
fGenesFDR_list_all_UpGene_vmedTDP_common <- 
  fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>% select(ENTREZID, major, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% 
  group_by(ENTREZID) %>% summarise(name_ct = n()) %>% filter(name_ct==2) %>% as.data.frame() #%>% nrow()
  fGenesFDR_list_all_UpGene_vmedTDP_common$name_ct <- gsub(2, "Up_vmedTDP_common", fGenesFDR_list_all_UpGene_vmedTDP_common$name_ct)
fGenesFDR_list_all_DownGene_vmedTDP_common <- 
  fGenesFDR_list_all %>% filter(grp=="vmedTDP") %>% select(ENTREZID, major, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% 
  group_by(ENTREZID) %>% summarise(name_ct = n()) %>% filter(name_ct==2) %>% as.data.frame() #%>% nrow()
  fGenesFDR_list_all_DownGene_vmedTDP_common$name_ct <- gsub(2, "Up_vmedTDP_common", fGenesFDR_list_all_DownGene_vmedTDP_common$name_ct)
##vnoTDP
fGenesFDR_list_all_UpGene_vnoTDP_common <- 
  fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>% select(ENTREZID, major, UpDown) %>% filter(UpDown=="UpGene") %>% unique() %>% 
  group_by(ENTREZID) %>% summarise(name_ct = n()) %>% filter(name_ct==2) %>% as.data.frame() #%>% nrow()
  fGenesFDR_list_all_UpGene_vnoTDP_common$name_ct <- gsub(2, "Up_vnoTDP_common", fGenesFDR_list_all_UpGene_vnoTDP_common$name_ct)
fGenesFDR_list_all_DownGene_vnoTDP_common <- 
  fGenesFDR_list_all %>% filter(grp=="vnoTDP") %>% select(ENTREZID, major, UpDown) %>% filter(UpDown=="DownGene") %>% unique() %>% 
  group_by(ENTREZID) %>% summarise(name_ct = n()) %>% filter(name_ct==2) %>% as.data.frame() #%>% nrow()
  fGenesFDR_list_all_DownGene_vnoTDP_common$name_ct <- gsub(2, "Up_vnoTDP_common", fGenesFDR_list_all_DownGene_vnoTDP_common$name_ct)  
  
  library("stringr")
  library("clusterProfiler")

## GO on the common 
fGene_common_TDPgrp_lst <- list(com_vhighTDP_Up=fGenesFDR_list_all_UpGene_vhighTDP_common, com_vhighTDP_Down=fGenesFDR_list_all_DownGene_vhighTDP_common,
                         com_vmedTDP_Up=fGenesFDR_list_all_UpGene_vmedTDP_common, com_vmedTDP_Down=fGenesFDR_list_all_DownGene_vmedTDP_common,
                         com_vnoTDP_Up=fGenesFDR_list_all_UpGene_vnoTDP_common, com_vnoTDP_Down=fGenesFDR_list_all_DownGene_vnoTDP_common
                         )
fGene_common_TDPgrp_lst_genes = lapply(fGene_common_TDPgrp_lst, function(i) as.data.frame(i)$ENTREZID)
  ## KEGG
fGene_common_lst_genes_TDPgrp_KEGG <- compareCluster(geneCluster   = fGene_common_TDPgrp_lst_genes,
                         fun           = "enrichKEGG",
                         pvalueCutoff  = 0.05, #keyType = "ENTREZID",
                         pAdjustMethod = "BH")
fGene_common_lst_genes_TDPgrp_goBP <- compareCluster(geneCluster   = fGene_common_TDPgrp_lst_genes, OrgDb=org.Hs.eg.db,ont = "BP",
                         fun           = "enrichGO",
                         pvalueCutoff  = 0.05, #keyType = "ENTREZID",
                         pAdjustMethod = "BH")
fGene_common_lst_genes_TDPgrp_goMF <- compareCluster(geneCluster   = fGene_common_TDPgrp_lst_genes, OrgDb=org.Hs.eg.db,ont = "MF",
                         fun           = "enrichGO",
                         pvalueCutoff  = 0.05, #keyType = "ENTREZID",
                         pAdjustMethod = "BH")

pdf(file="fGene_common_lst_genes_TDPgrp_GOs_plot_ALL.pdf", 
     width = 15, height = 20, onefile=TRUE)
dotplot(fGene_common_lst_genes_TDPgrp_KEGG, showCategory = 50, title = "KEGG Pathway Enrichment Analysis") +
  scale_y_discrete(labels=function(x) str_wrap(x, width=100))+theme(axis.text.x=element_text(angle=90,hjust=1))
dotplot(fGene_common_lst_genes_TDPgrp_goBP, showCategory = 50, title = "GO-biological processes Enrichment") + 
  scale_y_discrete(labels=function(x) str_wrap(x, width=100))+theme(axis.text.x=element_text(angle=90,hjust=1))
dotplot(fGene_common_lst_genes_TDPgrp_goMF, showCategory = 50, title = "GO-Molecular function Enrichment") + 
  scale_y_discrete(labels=function(x) str_wrap(x, width=100))+theme(axis.text.x=element_text(angle=90,hjust=1))
dotplot(gofilter(fGene_common_lst_genes_TDPgrp_goBP, level=4), showCategory = 50, title = "goBP_slim level 4") +
  scale_y_discrete(labels=function(x) str_wrap(x, width=100)) +theme(axis.text.x=element_text(angle=90,hjust=1))
dotplot(gofilter(fGene_common_lst_genes_TDPgrp_goBP, level=3), showCategory = 50, title = "goBP_slim level 3") +
  scale_y_discrete(labels=function(x) str_wrap(x, width=100))+theme(axis.text.x=element_text(angle=90,hjust=1))
dotplot(gofilter(fGene_common_lst_genes_TDPgrp_goBP, level=2), showCategory = 50, title = "goBP_slim level 2") +
  scale_y_discrete(labels=function(x) str_wrap(x, width=100))+theme(axis.text.x=element_text(angle=90,hjust=1))
dotplot(gofilter(fGene_common_lst_genes_TDPgrp_goMF, level=4), showCategory = 50, title = "goMF_slim level 4") +
  scale_y_discrete(labels=function(x) str_wrap(x, width=100))+theme(axis.text.x=element_text(angle=90,hjust=1))
dotplot(gofilter(fGene_common_lst_genes_TDPgrp_goMF, level=3), showCategory = 50, title = "goMF_slim level 3") +
  scale_y_discrete(labels=function(x) str_wrap(x, width=100))+theme(axis.text.x=element_text(angle=90,hjust=1))
dotplot(gofilter(fGene_common_lst_genes_TDPgrp_goMF, level=2), showCategory = 50, title = "goMF_slim level 2") +
  scale_y_discrete(labels=function(x) str_wrap(x, width=100))+theme(axis.text.x=element_text(angle=90,hjust=1))
dev.off()


library(forcats)
library(ggh4x)

## plot
pdf(file="fGene_common_lst_genes_TDPgrp_goMF_level3.pdf", 
     width = 3.5, height = 3.2)
gofilter(fGene_common_lst_genes_TDPgrp_goMF, level=3) %>% as.data.frame() %>% dplyr::select(Cluster, Description, qvalue) %>% 
  mutate(Cluster = factor(Cluster,levels=c("com_vhighTDP_Up", "com_vhighTDP_Down",
                                   "com_vmedTDP_Up","com_vmedTDP_Down",
                                   "com_vnoTDP_Up","com_vnoTDP_Down"))) %>%
  mutate(Description = fct_reorder(Description, -log10(qvalue))) %>% 
  ggplot(aes(x=-log10(qvalue), y=Description, fill=Cluster)) + facet_grid(rows = vars(Cluster), scales = "free", space = "free") + 
  geom_bar(stat="identity", color="black", position=position_dodge(), width = 0.85, size=0.2)+ scale_x_log10()+
  ylab("") + xlab(expression(-log[10]("FDR"))) + force_panelsizes(rows = c(1,3,1,10,2,5))+ geom_vline(xintercept = 2, colour="#ED1E79", linetype="dashed")+
  #scale_fill_manual(values=c('#D89000', '#15677A')) + 
  theme_Publication()
dev.off()  
    

# the setReadable didn't work somehow, why?
  write.table(setReadable(gofilter(fGene_common_lst_genes_TDPgrp_goMF, level=3), OrgDb = org.Hs.eg.db, keyType = "ENTREZID") %>% as.data.frame(),
              "fGene_common_lst_genes_TDPgrp_goMF_level3_df.txt", 
              append= F, sep='\t', 
              row.name = FALSE, col.names = TRUE, quote = FALSE)   

fGene_common_lst_genes_TDPgrp_goMF_df <- gofilter(fGene_common_lst_genes_TDPgrp_goMF, level=3) %>% as.data.frame()


for (U in 1:22) {
fGene_common_lst_genes_TDPgrp_goMF_df$geneID[U] <- 
  select(x = org.Hs.eg.db, keys = strsplit(fGene_common_lst_genes_TDPgrp_goMF_df$geneID[U], "\\/")[[1]], 
                                                          "SYMBOL", "ENTREZID")$SYMBOL %>% str_c(sep = "a", collapse ="/")
}

write.table(fGene_common_lst_genes_TDPgrp_goMF_df,
              "fGene_common_lst_genes_TDPgrp_goMF_level3_df.txt", 
              append= F, sep='\t', 
              row.name = FALSE, col.names = TRUE, quote = FALSE)  

#### get it out for the heatmap
common_ribo_lst <- data.frame() 
highTDP_riboGenes <- strsplit(fGene_common_lst_genes_TDPgrp_goMF_df$geneID[1], "\\/")[[1]]
highTDP_riboGenes_df <- data.frame(highTDP_riboGenes, "highTDP_up")
colnames(highTDP_riboGenes_df) <- c("SYMBOL", "enriched")

noTDP_riboGenes <-strsplit(fGene_common_lst_genes_TDPgrp_goMF_df$geneID[18], "\\/")[[1]]  
noTDP_riboGenes_df <- data.frame(noTDP_riboGenes, "noTDP_down")s
colnames(noTDP_riboGenes_df) <- c("SYMBOL", "enriched")


##medTDP
medTDP_riboGenes <-strsplit(fGene_common_lst_genes_TDPgrp_goMF_df$geneID[4] , "\\/")[[1]] 
medTDP_riboGenes_df <- data.frame(medTDP_riboGenes, "medTDP_up")
colnames(medTDP_riboGenes_df) <- c("SYMBOL", "enriched")

medTDP_riboGenes2 <-strsplit(fGene_common_lst_genes_TDPgrp_goMF_df$geneID[5] , "\\/")[[1]] 
medTDP_riboGenes_df2 <- data.frame(medTDP_riboGenes2, "medTDP_down")
colnames(medTDP_riboGenes_df2) <- c("SYMBOL", "enriched")


medTDP_riboGenes_df <- rbind(medTDP_riboGenes_df, medTDP_riboGenes_df2)
common_ribo_lst <- rbind(highTDP_riboGenes_df, noTDP_riboGenes_df, medTDP_riboGenes_df)

## then grab the data from the list, SYMBOL is used 

common_ribo_lst_geneFC <- merge(fGenesFDR_list_all, common_ribo_lst, by="SYMBOL", all.y = TRUE, incomparables ="NOMATCH") # %>% filter(grp!="vmedTDP")
common_ribo_lst_geneFC$Log2FC <- as.numeric(common_ribo_lst_geneFC$Log2FC) # <- as.numeric()

common_ribo_lst_geneFC_slim <- subset(common_ribo_lst_geneFC, select = c(SYMBOL,Log2FC, types))

## long to wide data frame
## should split it so there's no overlap 
common_ribo_lst_geneFC_highTDP <- 
  common_ribo_lst_geneFC %>% filter(enriched=="highTDP_down" | enriched=="highTDP_up") %>% filter(grp=="vhighTDP") %>% subset(select = c(SYMBOL,Log2FC, types)) %>% unique() %>% pivot_wider(names_from =types, values_from =Log2FC,values_fill = 0) %>% as.data.frame()
rownames(common_ribo_lst_geneFC_highTDP) <- common_ribo_lst_geneFC_highTDP$SYMBOL

common_ribo_lst_geneFC_noTDP <- 
  common_ribo_lst_geneFC %>% filter(enriched=="noTDP_down" | enriched=="noTDP_up") %>% filter(grp=="vnoTDP") %>% subset(select = c(SYMBOL,Log2FC, types)) %>% unique() %>% pivot_wider(names_from =types, values_from =Log2FC,values_fill = 0) %>% as.data.frame()
rownames(common_ribo_lst_geneFC_noTDP) <- common_ribo_lst_geneFC_noTDP$SYMBOL

common_ribo_lst_geneFC_medTDP <- 
  common_ribo_lst_geneFC %>% filter(enriched=="medTDP_down" | enriched=="medTDP_up") %>% filter(grp=="vmedTDP") %>% subset(select = c(SYMBOL,Log2FC, types)) %>% unique() %>% pivot_wider(names_from =types, values_from =Log2FC,values_fill = 0) %>% as.data.frame()
rownames(common_ribo_lst_geneFC_medTDP) <- common_ribo_lst_geneFC_medTDP$SYMBOL


rownames(common_ribo_lst_geneFC_wide) <- common_ribo_lst_geneFC_wide$SYMBOL
col_order <- c("C6_ODC_vhighTDP","C7_ODC_vhighTDP","C9_ODC_vhighTDP","C33_OPC_vhighTDP",
"C30_ASC_vhighTDP","C28_ASC_vhighTDP","C37_MG_vhighTDP","C2_MG_vhighTDP",
"C35_ENDO_vhighTDP","C14_EX_vhighTDP","C20_EX_vhighTDP","C17_EX_vhighTDP",
"C23_IN_vhighTDP","C13_IN_vhighTDP", "C7_ODC_vnoTDP","C6_ODC_vnoTDP","C9_ODC_vnoTDP","C32_OPC_vnoTDP","C33_OPC_vnoTDP",
"C30_ASC_vnoTDP","C28_ASC_vnoTDP","C37_MG_vnoTDP","C38_MG_vnoTDP","C2_MG_vnoTDP","C36_MG_vnoTDP","C35_ENDO_vnoTDP",
"C20_EX_vnoTDP" ,"C14_EX_vnoTDP","C12_EX_vnoTDP","C17_EX_vnoTDP","C21_EX_vnoTDP","C15_EX_vnoTDP",
"C26_IN_vnoTDP","C13_IN_vnoTDP","C23_IN_vnoTDP","C25_IN_vnoTDP","C22_IN_vnoTDP")

col_order_high <- c("C6_ODC_vhighTDP","C7_ODC_vhighTDP","C9_ODC_vhighTDP","C33_OPC_vhighTDP",
"C30_ASC_vhighTDP","C28_ASC_vhighTDP","C37_MG_vhighTDP","C2_MG_vhighTDP",
"C35_ENDO_vhighTDP","C14_EX_vhighTDP","C20_EX_vhighTDP","C17_EX_vhighTDP")

col_order_no <- c("C23_IN_vhighTDP","C13_IN_vhighTDP", "C7_ODC_vnoTDP","C6_ODC_vnoTDP","C9_ODC_vnoTDP","C32_OPC_vnoTDP","C33_OPC_vnoTDP",
"C30_ASC_vnoTDP","C28_ASC_vnoTDP","C37_MG_vnoTDP","C38_MG_vnoTDP","C2_MG_vnoTDP","C36_MG_vnoTDP","C35_ENDO_vnoTDP",
"C20_EX_vnoTDP" ,"C14_EX_vnoTDP","C12_EX_vnoTDP","C17_EX_vnoTDP","C21_EX_vnoTDP","C15_EX_vnoTDP",
"C26_IN_vnoTDP","C13_IN_vnoTDP","C23_IN_vnoTDP","C25_IN_vnoTDP","C22_IN_vnoTDP")

colnames(common_ribo_lst_geneFC_highTDP)
 [1] "SYMBOL"            "C20_EX_vhighTDP"   "C30_ASC_vhighTDP"  "C33_OPC_vhighTDP"  "C6_ODC_vhighTDP"   "C7_ODC_vhighTDP"   "C14_EX_vhighTDP"   "C2_MG_vhighTDP"   
 [9] "C28_ASC_vhighTDP"  "C9_ODC_vhighTDP"   "C35_ENDO_vhighTDP" "C37_MG_vhighTDP"   "C13_IN_vhighTDP"  
col_order_high <- c("C33_OPC_vhighTDP","C6_ODC_vhighTDP","C7_ODC_vhighTDP", "C9_ODC_vhighTDP",
"C30_ASC_vhighTDP","C28_ASC_vhighTDP",
"C2_MG_vhighTDP","C37_MG_vhighTDP",
"C35_ENDO_vhighTDP",
"C20_EX_vhighTDP","C14_EX_vhighTDP", "C13_IN_vhighTDP")  

col_order_no <- c("C32_OPC_vnoTDP", "C33_OPC_vnoTDP",
"C7_ODC_vnoTDP", "C6_ODC_vnoTDP","C9_ODC_vnoTDP",
"C30_ASC_vnoTDP","C28_ASC_vnoTDP",
"C37_MG_vnoTDP", "C38_MG_vnoTDP", "C2_MG_vnoTDP", "C36_MG_vnoTDP",
"C35_ENDO_vnoTDP",
"C20_EX_vnoTDP", "C14_EX_vnoTDP", "C12_EX_vnoTDP","C17_EX_vnoTDP",  "C21_EX_vnoTDP", "C15_EX_vnoTDP",
"C26_IN_vnoTDP", "C13_IN_vnoTDP",  "C23_IN_vnoTDP", "C25_IN_vnoTDP", "C22_IN_vnoTDP")

col_order_med <- c("C33_OPC_vmedTDP", "C32_OPC_vmedTDP",
                   "C7_ODC_vmedTDP",  "C6_ODC_vmedTDP", "C9_ODC_vmedTDP","C8_ODC_vmedTDP",
                   "C2_MG_vmedTDP", "C36_MG_vmedTDP",
                   "C12_EX_vmedTDP",  "C20_EX_vmedTDP","C15_EX_vmedTDP", "C21_EX_vmedTDP",  "C19_EX_vmedTDP","C14_EX_vmedTDP",
                   "C24_IN_vmedTDP", "C23_IN_vmedTDP",  "C26_IN_vmedTDP" ,  "C22_IN_vmedTDP" )

common_ribo_lst_geneFC_highTDP <- common_ribo_lst_geneFC_highTDP[,col_order_high] %>% unique()
common_ribo_lst_geneFC_noTDP <- common_ribo_lst_geneFC_wide[,col_order_no] %>% unique()
common_ribo_lst_geneFC_medTDP <- common_ribo_lst_geneFC_medTDP[,col_order_med] %>% unique()

common_ribo_lst_geneFC_highTDP_hm <- # this is too many, so hard to understand 
  Heatmap(as.matrix(common_ribo_lst_geneFC_highTDP), 
          name = "FC", show_row_names = TRUE,cluster_rows = TRUE, cluster_columns =FALSE, #row_km = 20,
          row_title_gp = gpar(fill = "#D9D0D3", fontsize=6), #row_title = c("1", "2", "3", "4", "5"), 
          row_title_rot = 0,
          col =col_ht, heatmap_legend_param = col_ht_lgd,
          row_gap = unit(0, "mm"), column_gap = unit(0, "mm"), border = TRUE, 
          #row_dend_reorder = TRUE, width = unit(1, "in"),
          column_names_rot = 90, column_names_side = "top", column_names_gp = gpar(fontsize = 6), column_names_centered = TRUE,
          row_names_gp = gpar(fontsize = 4))

common_ribo_lst_geneFC_noTDP_hm <- # this is too many, so hard to understand 
  Heatmap(as.matrix(common_ribo_lst_geneFC_noTDP), 
          name = "FC", show_row_names = TRUE,cluster_rows = TRUE, cluster_columns =FALSE, #row_km = 20,
          row_title_gp = gpar(fill = "#D9D0D3", fontsize=6), #row_title = c("1", "2", "3", "4", "5"), 
          row_title_rot = 0,
          col =col_ht, heatmap_legend_param = col_ht_lgd,
          row_gap = unit(0, "mm"), column_gap = unit(0, "mm"), border = TRUE, 
          #row_dend_reorder = TRUE, width = unit(1, "in"),
          column_names_rot = 90, column_names_side = "top", column_names_gp = gpar(fontsize = 6), column_names_centered = TRUE,
          row_names_gp = gpar(fontsize = 4))

common_ribo_lst_geneFC_medTDP_hm <- # this is too many, so hard to understand 
  Heatmap(as.matrix(common_ribo_lst_geneFC_medTDP), 
          name = "FC", show_row_names = TRUE,cluster_rows = TRUE, cluster_columns =FALSE, #row_km = 20,
          row_title_gp = gpar(fill = "#D9D0D3", fontsize=6), #row_title = c("1", "2", "3", "4", "5"), 
          row_title_rot = 0,
          col =col_ht, heatmap_legend_param = col_ht_lgd,
          row_gap = unit(0, "mm"), column_gap = unit(0, "mm"), border = TRUE, 
          #row_dend_reorder = TRUE, width = unit(1, "in"),
          column_names_rot = 90, column_names_side = "top", column_names_gp = gpar(fontsize = 6), column_names_centered = TRUE,
          row_names_gp = gpar(fontsize = 4))


pdf(file="common_ribo_lst_geneFC_highTDP_hm_small.pdf", 
    width = 4, height = 5, onefile = TRUE)
common_ribo_lst_geneFC_highTDP_hm #+ scale_y_discrete(limits=rev)
#+ scale_y_discrete(labels=function(x) str_wrap(x, width=100))
common_ribo_lst_geneFC_noTDP_hm
dev.off()    

pdf(file="common_ribo_lst_geneFC_medTDP.pdf", 
    width = 4, height = 5, onefile = TRUE)
common_ribo_lst_geneFC_medTDP_hm #+ scale_y_discrete(limits=rev)
dev.off()    

```

### C20 specific as shown in Figure 2H
```{r}

ArchR_C9[BiocGenerics::which(ArchR_C9$type2name =="EX")]
EX_markerGS <- getMarkerFeatures(
    ArchRProj = ArchR_C9[BiocGenerics::which(ArchR_C9$type2name =="EX")], 
    useMatrix = "GeneScoreMatrix", 
    groupBy = "type",
    bias = c("TSSEnrichment", "log10(nFrags)"),
    testMethod = "wilcoxon"
)

C20_markers <- 
 c( "VWA1", "LINC01770", "LINC00982", "CORT", "RCAN3", "MIR1976", "MIR5584", "GBP3", "RNVU1-19", "MIR4424", "BLACAT1", "MIR135B", "FAM89A", "MIR1182", "PRKCQ-AS1")
EX_markerGS_hm <- plotMarkerHeatmap(
  seMarker = EX_markerGS, 
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25", 
  #labelMarkers = type,
  transpose = FALSE, returnMatrix=FALSE)

# save PDF
plotPDF(EX_markerGS_hm, name = "GeneScores-Marker-Heatmap_EX_markerGS_hm", width = 4, height = 8, 
        ArchRProj = ArchR_C9, addDOC = FALSE)

ArchR_C9.combo_clu09.markersGS_hm <- plotMarkerHeatmap(
  seMarker = ArchR_C9.combo_clu09.markersGS, 
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25", 
  #labelMarkers = C20_markers,
  transpose = FALSE, returnMatrix=FALSE)

plotPDF(ArchR_C9.combo_clu09.markersGS_hm, name = "GeneScores-Marker-Heatmap_ArchR_C9.combo_clu09.markersGS_hm", width = 15, height = 15, 
        ArchRProj = ArchR_C9, addDOC = FALSE)

C20_NEU <-
	c("CYB561D1", "MIR4697HG", "MIR4697", "WNT10B", "WNT1", "MIR494", "MIR1193", "STRC", "C16orf92", "LPAR2", "IL13", "CAMK2A", "LINC00951", "FAM160B2", "MIR600HG")


C20_NEU_all <- c(C20_NEU,NEUgenes,EXgenes,L56genes)
  NEUgenes <- c("SNAP25", "SYT1") #neurons
  EXgenes <- c("SLC17A7", "SATB2", "RORB", "NEUROD2") #EX
  L56genes <- c("TLE4", "ASIC2", "KIAA1217", "MDGA2", "LRP1B")

ArchR_C9.combo_clu09.markersGS_2_hm <- plotMarkerHeatmap(
  seMarker = ArchR_C9.combo_clu09.markersGS, 
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25", 
  labelMarkers = C20_NEU_all,
  transpose = FALSE, returnMatrix=FALSE)
plotPDF(ArchR_C9.combo_clu09.markersGS_2_hm, name = "GeneScores-Marker-Heatmap_ArchR_C9.combo_clu09.markersGS_hm_2", width = 15, height = 15, 
         ArchRProj = ArchR_C9, addDOC = FALSE)
# make smaller for the figure 
plotPDF(ArchR_C9.combo_clu09.markersGS_2_hm, name = "GeneScores-Marker-Heatmap_ArchR_C9.combo_clu09.markersGS_hm_3", width = 6, height = 5, 
        ArchRProj = ArchR_C9, addDOC = FALSE)  

colnames(ArchR_C9.combo_clu09.markersGS)=="C20"
```

### mitochrondira plot for all clusters in disease/TDP-43 level specific enrichments
```{r}
### check the GO-term related to mitochrondria nad 
fGenes_TDPlev_Down_KEGG_df %>% arrange(Description) %>% group_by(Description) %>% mutate(Desp_n=n()) %>% ungroup() %>% select(Description, Desp_n) %>% unique() %>% arrange(-Desp_n)

mito <- c("Chemical carcinogenesis - reactive oxygen species", "Oxidative phosphorylation", "Mitophagy - animal", "Autophagy - animal", # KEGG
  #"axonogenesis",
"aerobic respiration",
"ATP metabolic process"	,
"oxidative phosphorylation"	,
"mitochondrial electron transport, NADH to ubiquinone",
"mitochondrial respiratory chain complex I assembly",
"mitophagy",
"autophagy of mitochondrion",
"macroautophagy", # GOBP
"NADH dehydrogenase activity" # GOMG
)

fGenes_TDPlev_Down_GOBP_df %>% arrange(Description) %>% group_by(Description) %>% mutate(Desp_n=n()) %>% ungroup() %>% select(Description, Desp_n) %>% unique() %>% arrange(-Desp_n)
fGenes_TDPlev_Down_GOMF_df %>% arrange(Description) %>% group_by(Description) %>% mutate(Desp_n=n()) %>% ungroup() %>% select(Description, Desp_n) %>% unique() %>% arrange(-Desp_n)

### make one single grph  

GeneRatio <- mito_df$GeneRatio
mito_df <- subset(rbind(fGenes_TDPlev_Down_KEGG_df, fGenes_TDPlev_Down_GOBP_df, fGenes_TDPlev_Down_GOMF_df), Description %in% mito)
GeneRatio <- mito_df$GeneRatio

mito_df$GeneRatio <- sapply(strsplit(GeneRatio, "/"), function(x) x[1]) %>% as.numeric() /sapply(strsplit(GeneRatio, "/"), function(x) x[2])%>% as.numeric() 

## order eveything 

mito_df <- mito_df %>% mutate(Description = factor(Description, levels=rev(mito)))

write.table(rbind(mito_df, ASC_m,ENDO_highTDP, ENDO_medTDP,OPC_highTDP),
              "mito_df.csv", 
              append= F, sep=',', 
              row.name = FALSE, col.names = TRUE, quote = FALSE)  

pdf(file="Mito_GO_plot.pdf", 
     width = 3.4, height = 1.6, onefile=TRUE)
  rbind(mito_df, ASC_m,ENDO_highTDP, ENDO_medTDP,OPC_highTDP) %>%
          ggplot(aes(x = Cluster, y = Description)) + 
             geom_point(aes(size = GeneRatio, color = p.adjust)) +
             theme_bw(base_size = 4) + scale_size_continuous(range = c(0.1,2))+
             scale_colour_gradient(limits=c(0, 0.05), low="red", high="blue") +
             ylab(NULL) +
             ggtitle("mitochondria/respiration GO-enrichment") +
      theme_prism(base_size = 4, border = TRUE, base_fontface = "plain") + theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))

dev.off() 

## fill the missing ones
ASC_m <- mito_df[1,]
    ASC_m[1,1] <- "ASC_medTDP"
    ASC_m[1,7] <- 0.05
ENDO_highTDP <- mito_df[1,]
    ENDO_highTDP[1,1] <- "ENDO_highTDP"
    ENDO_highTDP[1,7] <- 0.05
 ENDO_medTDP <- mito_df[1,]
    ENDO_medTDP[1,1] <- "ENDO_medTDP"
    ENDO_medTDP[1,7] <- 0.05   
 OPC_highTDP <- mito_df[1,]
    OPC_highTDP[1,1] <- "OPC_highTDP"
    OPC_highTDP[1,7] <- 0.05       
    

### find the specific one     
mito_df2 <- rbind(mito_df, ASC_m,ENDO_highTDP, ENDO_medTDP,OPC_highTDP)

mito_df2$type <- sapply(strsplit(mito_df2$Cluster %>% as.character() , "_"), function(x) x[1])
mito_df2$DEG <- sapply(strsplit(mito_df2$Cluster %>% as.character() , "_"), function(x) x[2])

mito_comxI <- mito_df2 %>% filter(Description=="mitochondrial respiratory chain complex I assembly") %>% filter(type == "ODC" | type=="EX")
```

## Differential chromatin accessibility analysis and enrichment of TF motifs in differentially accessible regions
```{r}
### use the same scripts, expecept for change useMatrix = "PeakMatrix", in the getMarkerFeatures() as described in the differential comparison portion
# see detials below

### do the peaks, for all 35 
  ## start 1, 2:3, 4:5
for (U in 1:35) {
  print(paste("START with--U --",U, sep=""))
    TYPE <- gsub("-", "_",clu38TDP_compo$type[U])

  print(paste("START with U --",U, ":: TYPE is ", TYPE, sep=""))
      CTL <- clu38TDP_compo$control[U]
      highTDP <- clu38TDP_compo$highTDP[U]
      medTDP <- clu38TDP_compo$medTDP[U]
      noTDP <- clu38TDP_compo$noTDP[U]
  print(paste("compare::", CTL, "with:", highTDP, medTDP, noTDP, sep=' '))
  
  print(paste(TYPE,"-- STEP 1 -- fGene comp",sep=""))
print("=== v highTDP")
fPeaks_vhighTDP <- getMarkerFeatures(
  ArchRProj = ArchR_C9, 
  useMatrix = "PeakMatrix",
  groupBy = "clu38TDP",testMethod = "wilcoxon",bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = highTDP,bgdGroups = CTL)
print("=== v medTDP")
fPeaks_vmedTDP <- getMarkerFeatures(
  ArchRProj = ArchR_C9, 
  useMatrix = "PeakMatrix",
  groupBy = "clu38TDP",testMethod = "wilcoxon",bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = medTDP,bgdGroups = CTL)
print("=== v noTDP")
fPeaks_vnoTDP <- getMarkerFeatures(
  ArchRProj = ArchR_C9, 
  useMatrix = "PeakMatrix",
  groupBy = "clu38TDP",testMethod = "wilcoxon",bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = noTDP,bgdGroups = CTL)
# save out to a name 
assign(paste("fPeaks_",TYPE,"_vhighTDP",sep=""), fPeaks_vhighTDP)
assign(paste("fPeaks_",TYPE,"_vmedTDP",sep=""), fPeaks_vmedTDP)
assign(paste("fPeaks_",TYPE,"_vnoTDP",sep=""), fPeaks_vnoTDP)

## save RDS
print("== SAVE RDS")
  saveRDS(fPeaks_vhighTDP, paste(fPeaks_dir,paste("fPeaks_",TYPE,"_vhighTDP",sep=""),".rds",sep="")) 
  saveRDS(fPeaks_vmedTDP, paste(fPeaks_dir,paste("fPeaks_",TYPE,"_vmedTDP",sep=""),".rds",sep="")) 
  saveRDS(fPeaks_vnoTDP, paste(fPeaks_dir,paste("fPeaks_",TYPE,"_vnoTDP",sep=""),".rds",sep="")) 

# filter: 
print("== Filter and write out") # the first item on the list 
fPeaks_vhighTDP_FDR <- getMarkers(fPeaks_vhighTDP, cutOff = "FDR <= 0.05 & abs(Log2FC) > 0.5")
  write.table(fPeaks_vhighTDP_FDR[[1]],paste(fPeaks_dirFDR,paste("fPeaks_",TYPE,"_vhighTDP",sep=""),".FDR.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
fPeaks_vmedTDP_FDR <- getMarkers(fPeaks_vmedTDP, cutOff = "FDR <= 0.05 & abs(Log2FC) > 0.5")
  write.table(fPeaks_vmedTDP_FDR[[1]],paste(fPeaks_dirFDR,paste("fPeaks_",TYPE,"_vmedTDP",sep=""),".FDR.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
fPeaks_vnoTDP_FDR <- getMarkers(fPeaks_vnoTDP, cutOff = "FDR <= 0.05 & abs(Log2FC) > 0.5")
  write.table(fPeaks_vnoTDP_FDR[[1]],paste(fPeaks_dirFDR,paste("fPeaks_",TYPE,"_vnoTDP",sep=""),".FDR.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
    # save out to a name 
    assign(paste("fPeaks_",TYPE,"_vhighTDP","_FDR",sep=""), fPeaks_vhighTDP_FDR)
    assign(paste("fPeaks_",TYPE,"_vmedTDP","_FDR",sep=""), fPeaks_vmedTDP_FDR)
    assign(paste("fPeaks_",TYPE,"_vnoTDP","_FDR",sep=""), fPeaks_vnoTDP_FDR)

## get the whole list out 
print("== write out whole list") # the first item on the list 
fPeaks_vhighTDP_noflt <- getMarkers(fPeaks_vhighTDP, cutOff = "abs(Log2FC)>=0")
  fPeaks_vhighTDP_noflt.df <- getMarkers(fPeaks_vhighTDP, cutOff = "abs(Log2FC)>=0")[[1]] %>% as.data.frame()
  fPeaks_vhighTDP_noflt.df$ID <- paste(fPeaks_vhighTDP_noflt.df$seqnames, fPeaks_vhighTDP_noflt.df$start,fPeaks_vhighTDP_noflt.df$end, sep="_")
  fPeaks_vhighTDP_noflt.df2 <- merge(fPeaks_vhighTDP_noflt.df, peakMaster, by.x="ID", by.y="seqnames_start_end")
    write.table(fPeaks_vhighTDP_noflt.df2,paste(fPeaks_dirnoflt,paste("fPeaks_",TYPE,"_vhighTDP",sep=""),".noflt.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
fPeaks_vmedTDP_noflt <- getMarkers(fPeaks_vmedTDP, cutOff = "abs(Log2FC)>=0")
  fPeaks_vmedTDP_noflt.df <- getMarkers(fPeaks_vmedTDP, cutOff = "abs(Log2FC)>=0")[[1]] %>% as.data.frame()
  fPeaks_vmedTDP_noflt.df$ID <- paste(fPeaks_vmedTDP_noflt.df$seqnames, fPeaks_vmedTDP_noflt.df$start,fPeaks_vmedTDP_noflt.df$end, sep="_")
  fPeaks_vmedTDP_noflt.df2 <- merge(fPeaks_vmedTDP_noflt.df, peakMaster, by.x="ID", by.y="seqnames_start_end")
    write.table(fPeaks_vmedTDP_noflt.df2,paste(fPeaks_dirnoflt,paste("fPeaks_",TYPE,"_vmedTDP",sep=""),".noflt.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
fPeaks_vnoTDP_noflt <- getMarkers(fPeaks_vnoTDP, cutOff = "abs(Log2FC)>=0")
  fPeaks_vnoTDP_noflt.df <- getMarkers(fPeaks_vnoTDP, cutOff = "abs(Log2FC)>=0")[[1]] %>% as.data.frame()
  fPeaks_vnoTDP_noflt.df$ID <- paste(fPeaks_vnoTDP_noflt.df$seqnames, fPeaks_vnoTDP_noflt.df$start,fPeaks_vnoTDP_noflt.df$end, sep="_")
  fPeaks_vnoTDP_noflt.df2 <- merge(fPeaks_vnoTDP_noflt.df, peakMaster, by.x="ID", by.y="seqnames_start_end")
    write.table(fPeaks_vnoTDP_noflt.df2,paste(fPeaks_dirnoflt,paste("fPeaks_",TYPE,"_vnoTDP",sep=""),".noflt.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)

## plot MA
print("== PLOT MA -- donesn't work, so skip -- see old fPEaks")
######  peak info mapping:: start ArchR_C9_PeakGR
print("== peak info mapping:: start ArchR_C9_PeakGR, skipped -- see old fPEaks")

##################  
    
##### peak-motif enrichment --- ##### filter:   cutOff = "FDR <= 0.1 & Log2FC >= 0.5", I use FC >1 = grt 2 fold 
print("== peak motif enrichment:: Up and down:: FDR 0.1 | abs(log2FC)> 1")
##### vhighTDP ########### 
print("== set 1 == vhighTDP")
fPeaks_vhighTDP_Up <- peakAnnoEnrichment(
    seMarker = fPeaks_vhighTDP,
    ArchRProj = ArchR_C9, peakAnnotation = "Motif",background = "bgdPeaks",
    cutOff = "FDR <= 0.1 & Log2FC >= 1")
fPeaks_vhighTDP_Down <- peakAnnoEnrichment(
    seMarker = fPeaks_vhighTDP,
    ArchRProj = ArchR_C9, peakAnnotation = "Motif",background = "bgdPeaks",
    cutOff = "FDR <= 0.1 & Log2FC <= -1")
fPeaks_vhighTDP_Up_df <- data.frame(TF = rownames(fPeaks_vhighTDP_Up), 
                         mlog10Padj = assay(fPeaks_vhighTDP_Up)[,1], 
                         X=assays(fPeaks_vhighTDP_Up)$CompareFrequency)
        fPeaks_vhighTDP_Up_df <- fPeaks_vhighTDP_Up_df[order(fPeaks_vhighTDP_Up_df$mlog10Padj, decreasing = TRUE),]
        fPeaks_vhighTDP_Up_df$rank <- seq_len(nrow(fPeaks_vhighTDP_Up_df))
fPeaks_vhighTDP_Down_df <- data.frame(TF = rownames(fPeaks_vhighTDP_Down), 
                         mlog10Padj = assay(fPeaks_vhighTDP_Down)[,1], 
                         X=assays(fPeaks_vhighTDP_Down)$CompareFrequency)
        fPeaks_vhighTDP_Down_df <- fPeaks_vhighTDP_Down_df[order(fPeaks_vhighTDP_Down_df$mlog10Padj, decreasing = TRUE),]
        fPeaks_vhighTDP_Down_df$rank <- seq_len(nrow(fPeaks_vhighTDP_Down_df))
write.table(fPeaks_vhighTDP_Up_df,paste(fPeaks_dirFDR_motifs,paste("fPeaks_",TYPE,"_vhighTDP",sep=""),".FDR.Up_enMotifs.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
write.table(fPeaks_vhighTDP_Down_df,paste(fPeaks_dirFDR_motifs,paste("fPeaks_",TYPE,"_vhighTDP",sep=""),".FDR.Down_enMotifs.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)

### plotting the enrich motifs fPeaks_dirFDR_motifsPlot
print("== plotting the enrich motifs")    
ggUp <- ggplot(fPeaks_vhighTDP_Up_df, aes(rank, mlog10Padj, color = mlog10Padj)) + 
  geom_point(size = 1) + ggrepel::geom_label_repel(
        data = fPeaks_vhighTDP_Up_df[rev(seq_len(30)), ], aes(x = rank, y = mlog10Padj, label = TF), 
        size = 1.5,nudge_x = 2,color = "black") + theme_ArchR() + 
  ylab("-log10(P-adj) Motif Enrichment") + xlab(paste("UP:TFmotif_EN", TYPE,"_vhighTDP", sep="")) +
  scale_color_gradientn(colors = paletteContinuous(set = "comet"))
ggDown<- ggplot(fPeaks_vhighTDP_Down_df, aes(rank, mlog10Padj, color = mlog10Padj)) + 
  geom_point(size = 1) + ggrepel::geom_label_repel(
        data = fPeaks_vhighTDP_Down_df[rev(seq_len(30)), ], aes(x = rank, y = mlog10Padj, label = TF), 
        size = 1.5,nudge_x = 2,color = "black") + theme_ArchR() + 
  ylab("-log10(P-adj) Motif Enrichment") + xlab(paste("Down:TFmotif_EN", TYPE,"_vhighTDP", sep="")) +
  scale_color_gradientn(colors = paletteContinuous(set = "comet"))
pdf(file=paste(fPeaks_dirFDR_motifsPlot,paste("fPeaks_",TYPE,"_vhighTDP",sep=""),".FDR.motifs.pdf",sep=""), width = 5, height = 5)
  for(k in 1) {
    print(ggUp)
    print(ggDown)}
dev.off()
##### END vhighTDP ####
##### vmedTDP ########### 
print("== set 1 == vmedTDP")
fPeaks_vmedTDP_Up <- peakAnnoEnrichment(
    seMarker = fPeaks_vmedTDP,
    ArchRProj = ArchR_C9, peakAnnotation = "Motif",background = "bgdPeaks",
    cutOff = "FDR <= 0.1 & Log2FC >= 1")
fPeaks_vmedTDP_Down <- peakAnnoEnrichment(
    seMarker = fPeaks_vmedTDP,
    ArchRProj = ArchR_C9, peakAnnotation = "Motif",background = "bgdPeaks",
    cutOff = "FDR <= 0.1 & Log2FC <= -1")
fPeaks_vmedTDP_Up_df <- data.frame(TF = rownames(fPeaks_vmedTDP_Up), 
                         mlog10Padj = assay(fPeaks_vmedTDP_Up)[,1], 
                         X=assays(fPeaks_vmedTDP_Up)$CompareFrequency)
        fPeaks_vmedTDP_Up_df <- fPeaks_vmedTDP_Up_df[order(fPeaks_vmedTDP_Up_df$mlog10Padj, decreasing = TRUE),]
        fPeaks_vmedTDP_Up_df$rank <- seq_len(nrow(fPeaks_vmedTDP_Up_df))
fPeaks_vmedTDP_Down_df <- data.frame(TF = rownames(fPeaks_vmedTDP_Down), 
                         mlog10Padj = assay(fPeaks_vmedTDP_Down)[,1], 
                         X=assays(fPeaks_vmedTDP_Down)$CompareFrequency)
        fPeaks_vmedTDP_Down_df <- fPeaks_vmedTDP_Down_df[order(fPeaks_vmedTDP_Down_df$mlog10Padj, decreasing = TRUE),]
        fPeaks_vmedTDP_Down_df$rank <- seq_len(nrow(fPeaks_vmedTDP_Down_df))
write.table(fPeaks_vmedTDP_Up_df,paste(fPeaks_dirFDR_motifs,paste("fPeaks_",TYPE,"_vmedTDP",sep=""),".FDR.Up_enMotifs.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
write.table(fPeaks_vmedTDP_Down_df,paste(fPeaks_dirFDR_motifs,paste("fPeaks_",TYPE,"_vmedTDP",sep=""),".FDR.Down_enMotifs.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)

### plotting the enrich motifs fPeaks_dirFDR_motifsPlot
print("== plotting the enrich motifs")    
ggUp <- ggplot(fPeaks_vmedTDP_Up_df, aes(rank, mlog10Padj, color = mlog10Padj)) + 
  geom_point(size = 1) + ggrepel::geom_label_repel(
        data = fPeaks_vmedTDP_Up_df[rev(seq_len(30)), ], aes(x = rank, y = mlog10Padj, label = TF), 
        size = 1.5,nudge_x = 2,color = "black") + theme_ArchR() + 
  ylab("-log10(P-adj) Motif Enrichment") + xlab(paste("UP:TFmotif_EN", TYPE,"_vmedTDP", sep="")) +
  scale_color_gradientn(colors = paletteContinuous(set = "comet"))
ggDown<- ggplot(fPeaks_vmedTDP_Down_df, aes(rank, mlog10Padj, color = mlog10Padj)) + 
  geom_point(size = 1) + ggrepel::geom_label_repel(
        data = fPeaks_vmedTDP_Down_df[rev(seq_len(30)), ], aes(x = rank, y = mlog10Padj, label = TF), 
        size = 1.5,nudge_x = 2,color = "black") + theme_ArchR() + 
  ylab("-log10(P-adj) Motif Enrichment") + xlab(paste("Down:TFmotif_EN", TYPE,"_vmedTDP", sep="")) +
  scale_color_gradientn(colors = paletteContinuous(set = "comet"))
pdf(file=paste(fPeaks_dirFDR_motifsPlot,paste("fPeaks_",TYPE,"_vmedTDP",sep=""),".FDR.motifs.pdf",sep=""), width = 5, height = 5)
  for(k in 1) {
    print(ggUp)
    print(ggDown)}
dev.off()
##### END vmedTDP ####
##### vnoTDP ########### 
print("== set 1 == vnoTDP")
fPeaks_vnoTDP_Up <- peakAnnoEnrichment(
    seMarker = fPeaks_vnoTDP,
    ArchRProj = ArchR_C9, peakAnnotation = "Motif",background = "bgdPeaks",
    cutOff = "FDR <= 0.1 & Log2FC >= 1")
fPeaks_vnoTDP_Down <- peakAnnoEnrichment(
    seMarker = fPeaks_vnoTDP,
    ArchRProj = ArchR_C9, peakAnnotation = "Motif",background = "bgdPeaks",
    cutOff = "FDR <= 0.1 & Log2FC <= -1")
fPeaks_vnoTDP_Up_df <- data.frame(TF = rownames(fPeaks_vnoTDP_Up), 
                         mlog10Padj = assay(fPeaks_vnoTDP_Up)[,1], 
                         X=assays(fPeaks_vnoTDP_Up)$CompareFrequency)
        fPeaks_vnoTDP_Up_df <- fPeaks_vnoTDP_Up_df[order(fPeaks_vnoTDP_Up_df$mlog10Padj, decreasing = TRUE),]
        fPeaks_vnoTDP_Up_df$rank <- seq_len(nrow(fPeaks_vnoTDP_Up_df))
fPeaks_vnoTDP_Down_df <- data.frame(TF = rownames(fPeaks_vnoTDP_Down), 
                         mlog10Padj = assay(fPeaks_vnoTDP_Down)[,1], 
                         X=assays(fPeaks_vnoTDP_Down)$CompareFrequency)
        fPeaks_vnoTDP_Down_df <- fPeaks_vnoTDP_Down_df[order(fPeaks_vnoTDP_Down_df$mlog10Padj, decreasing = TRUE),]
        fPeaks_vnoTDP_Down_df$rank <- seq_len(nrow(fPeaks_vnoTDP_Down_df))
write.table(fPeaks_vnoTDP_Up_df,paste(fPeaks_dirFDR_motifs,paste("fPeaks_",TYPE,"_vnoTDP",sep=""),".FDR.Up_enMotifs.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)
write.table(fPeaks_vnoTDP_Down_df,paste(fPeaks_dirFDR_motifs,paste("fPeaks_",TYPE,"_vnoTDP",sep=""),".FDR.Down_enMotifs.csv",sep=""), 
          append= F, sep=',', 
          row.name = FALSE, col.names = TRUE, quote = FALSE)

### plotting the enrich motifs fPeaks_dirFDR_motifsPlot
print("== plotting the enrich motifs")    
ggUp <- ggplot(fPeaks_vnoTDP_Up_df, aes(rank, mlog10Padj, color = mlog10Padj)) + 
  geom_point(size = 1) + ggrepel::geom_label_repel(
        data = fPeaks_vnoTDP_Up_df[rev(seq_len(30)), ], aes(x = rank, y = mlog10Padj, label = TF), 
        size = 1.5,nudge_x = 2,color = "black") + theme_ArchR() + 
  ylab("-log10(P-adj) Motif Enrichment") + xlab(paste("UP:TFmotif_EN", TYPE,"_vnoTDP", sep="")) +
  scale_color_gradientn(colors = paletteContinuous(set = "comet"))
ggDown<- ggplot(fPeaks_vnoTDP_Down_df, aes(rank, mlog10Padj, color = mlog10Padj)) + 
  geom_point(size = 1) + ggrepel::geom_label_repel(
        data = fPeaks_vnoTDP_Down_df[rev(seq_len(30)), ], aes(x = rank, y = mlog10Padj, label = TF), 
        size = 1.5,nudge_x = 2,color = "black") + theme_ArchR() + 
  ylab("-log10(P-adj) Motif Enrichment") + xlab(paste("Down:TFmotif_EN", TYPE,"_vnoTDP", sep="")) +
  scale_color_gradientn(colors = paletteContinuous(set = "comet"))
pdf(file=paste(fPeaks_dirFDR_motifsPlot,paste("fPeaks_",TYPE,"_vnoTDP",sep=""),".FDR.motifs.pdf",sep=""), width = 5, height = 5)
  for(k in 1) {
    print(ggUp)
    print(ggDown)}
dev.off()
##### END vnoTDP ####
}
```     

